// @triplyetl/vocabularies/lib/Vocabularies/Static/cc.js
/* eslint-disable quote-props */
import VocabularyBase from '../../Classes/Vocabulary.js';
class Licence extends VocabularyBase {
    '1.0' = this.concat('1.0/');
    '2.0' = this.concat('2.0/');
    '2.5' = this.concat('2.5/');
    '3.0' = this.concat('3.0/');
    '4.0' = this.concat('4.0/');
}
class by extends Licence {
}
class by_nc extends Licence {
}
class by_nc_nd extends Licence {
}
class by_nc_sa extends Licence {
}
class by_nd extends Licence {
}
class by_sa extends Licence {
}
class zero extends VocabularyBase {
    '1.0' = this.concat('1.0/');
}
class Vocabulary extends VocabularyBase {
    by = new by('https://creativecommons.org/licenses/by/');
    'by-nc' = new by_nc('https://creativecommons.org/licenses/by-nc/');
    'by-nc-nd' = new by_nc_nd('https://creativecommons.org/licenses/by-nc-nd/');
    'by-nc-sa' = new by_nc_sa('https://creativecommons.org/licenses/by-nc-sa/');
    'by-nd' = new by_nd('https://creativecommons.org/licenses/by-nd/');
    'by-sa' = new by_sa('https://creativecommons.org/licenses/by-sa/');
    'zero' = new zero('https://creativecommons.org/publicdomain/zero/');
}
/**
 * prefix **cc:** <<http://creativecommons.org/ns#>>
 *
 * Creative Commons licenses.
 */
const cc = new Vocabulary('http://creativecommons.org/ns#');
export default cc;
