// @triplyetl/etl/src/ratt/__tests__/getEtlForTesting.ts
import { NamedNode } from "@triplydb/data-factory/Terms.js";
import { Iri, Middleware, MiddlewareList } from "../../generic/index.js";
import Etl from "../../generic/Etl.js";
import { fromJson } from "../../generic/middlewares/sources/index.js";

export function getEtlForTesting(
  middlewares?: MiddlewareList | Middleware,
  _prefixes?: { [key: string]: (l: string) => NamedNode<string> },
): Etl {
  const prefixes = {
    ..._prefixes,
    ex: Iri("http://ex.com/"),
    blanknode: Iri("http://ex.com/.well-known/genid/"),
    xsd: Iri("http://www.w3.org/2001/XMLSchema#"),
  };
  const etl = new Etl({ prefixes });
  etl.use(fromJson({ foo: "bar" }));
  if (middlewares !== undefined) etl.use(middlewares);
  return etl;
}

export function getEtlForTestingNoExtractor(
  middlewares?: MiddlewareList | Middleware,
  _prefixes?: { [key: string]: (l: string) => NamedNode<string> },
): Etl {
  const prefixes = {
    ..._prefixes,
    ex: Iri("http://ex.com/"),
    blanknode: Iri("http://ex.com/.well-known/genid/"),
    xsd: Iri("http://www.w3.org/2001/XMLSchema#"),
  };
  const etl = new Etl({ prefixes });
  if (middlewares !== undefined) etl.use(middlewares);
  return etl;
}
