// @triplyetl/etl/src/utils/HttpRequestHeaders.ts
// Sourced from https://en.wikipedia.org/wiki/List_of_HTTP_header_fields

interface HttpRequestHeaders {
  /**
   * Acceptable instance-manipulations for the request.
   * @example A-IM: feed
   * @see https://datatracker.ietf.org/doc/html/rfc3229
   */
  "A-IM"?: string;

  /**
   * Media type(s) that is/are acceptable for the response. See Content negotiation.
   * @example Accept: text/html
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  Accept?: string;

  /**
   * Character sets that are acceptable.
   * @example Accept-Charset: utf-8
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  "Accept-Charset"?: string;

  /**
   * Acceptable version in time.
   * @example Accept-Datetime: Thu, 31 May 2007 20:35:00 GMT
   * @see https://datatracker.ietf.org/doc/html/rfc7089
   */
  "Accept-Datetime"?: string;

  /**
   * List of acceptable encodings. See HTTP compression.
   * @example Accept-Encoding: gzip, deflate
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  "Accept-Encoding"?: string;

  /**
   * List of acceptable human languages for response. See Content negotiation.
   * @example Accept-Language: en-US
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  "Accept-Language"?: string;

  /**
   * Initiates a request for cross-origin resource sharing with Origin (below).
   * @example Access-Control-Request-Method: GET
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  "Access-Control-Request-Method"?: string;

  /**
   * Authentication credentials for HTTP authentication.
   *
   * @example Authorization: Basic QWxhZGRpbjpvcGVuIHNlc2FtZQ==
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  Authorization?: string;

  /**
   * Used to specify directives that must be obeyed by all caching mechanisms along the request-response chain.
   * @example Cache-Control: no-cache
   * @see https://datatracker.ietf.org/doc/html/rfc9111
   */
  "Cache-Control"?: string;

  /**
   * Control options for the current connection and list of hop-by-hop request fields.
   * Must not be used with HTTP/2.
   * @example Connection: keep-alive
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  Connection?: string;

  /**
   * The type of encoding used on the data. See HTTP compression.
   * @example Content-Encoding: gzip
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  "Content-Encoding"?: string;

  /**
   * The length of the request body in octets (8-bit bytes).
   * @example Content-Length: 348
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  "Content-Length"?: string;

  /**
   * A Base64-encoded binary MD5 sum of the content of the request body.
   * @example Content-MD5: Q2hlY2sgSW50ZWdyaXR5IQ==
   * @see https://datatracker.ietf.org/doc/html/rfc1544, 1864, 4021
   */
  "Content-MD5"?: string;

  /**
   * The Media type of the body of the request (used with POST and PUT requests).
   * @example Content-Type: application/x-www-form-urlencoded
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  "Content-Type"?: string;

  /**
   * An HTTP cookie previously sent by the server with Set-Cookie (below).
   * @example Cookie: $Version=1; Skin=new;
   * @see https://datatracker.ietf.org/doc/html/rfc2965, 6265
   */
  Cookie?: string;

  /**
   * The date and time at which the message was originated
   * (in "HTTP-date" format as defined by RFC 9110:
   * HTTP Semantics, section 5.6.7 "Date/Time Formats").
   * @example Date: Tue, 15 Nov 1994 08:12:31 GMT
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  Date?: string;

  /**
   * Indicates that particular server behaviors are required by the client.
   * @example Expect: 100-continue
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  Expect?: string;

  /**
   * Disclose original information of a client connecting to a web server through an HTTP proxy.
   * @example Forwarded: for=192.0.2.60;proto=http;by=203.0.113.43 Forwarded: for=192.0.2.43, for=198.51.100.17
   * @see https://datatracker.ietf.org/doc/html/rfc7239
   */
  Forwarded?: string;

  /**
   * The email address of the user making the request.
   * @example From: user@example.com
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  From?: string;

  /**
   * The domain name of the server (for virtual hosting), and the TCP port number on which the server is listening.
   * The port number may be omitted if the port is the standard port for the service requested.
   * Mandatory since HTTP/1.1. If the request is generated directly in HTTP/2, it should not be used.
   * @example Host: en.wikipedia.org
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  Host?: string;

  /**
   * A request that upgrades from HTTP/1.1 to HTTP/2 MUST include exactly one HTTP2-Setting header field.
   * The HTTP2-Settings header field is a connection-specific header field that includes parameters that govern the HTTP/2 connection, provided in anticipation of the server accepting the request to upgrade.
   * @example HTTP2-Settings: token64
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  "HTTP2-Settings"?: string;

  /**
   * Only perform the action if the client supplied entity matches the same entity on the server.
   * This is mainly for methods like PUT to only update a resource if it has not been modified since the user last updated it.
   * @example If-Match: "737060cd8c284d8af7ad3082f209582d"
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  "If-Match"?: string;

  /**
   * Allows a 304 Not Modified to be returned if content is unchanged.
   * @example If-Modified-Since: Sat, 29 Oct 1994 19:43:31 GMT
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  "If-Modified-Since"?: string;

  /**
   * Allows a 304 Not Modified to be returned if content is unchanged, see HTTP ETag.
   * @example If-None-Match: "737060cd8c284d8af7ad3082f209582d"
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  "If-None-Match"?: string;

  /**
   * If the entity is unchanged, send me the part(s) that I am missing; otherwise, send me the entire new entity.
   * @example If-Range: "737060cd8c284d8af7ad3082f209582d"
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  "If-Range"?: string;

  /**
   * Only send the response if the entity has not been modified since a specific time.
   * @example If-Unmodified-Since: Sat, 29 Oct 1994 19:43:31 GMT
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  "If-Unmodified-Since"?: string;

  /**
   * Limit the number of times the message can be forwarded through proxies or gateways.
   * @example Max-Forwards: 10
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  "Max-Forwards"?: string;

  /**
   * Initiates a request for cross-origin resource sharing (asks server for Access-Control-* response fields).
   * @example Origin: http://www.example-social-network.com
   * @see https://datatracker.ietf.org/doc/html/rfc6454
   */
  Origin?: string;

  /**
   * Implementation-specific fields that may have various effects anywhere along the request-response chain.
   * @example Pragma: no-cache
   * @see https://datatracker.ietf.org/doc/html/rfc9111
   */
  Pragma?: string;

  /**
   * Allows client to request that certain behaviors be employed by a server while processing a request.
   * @example Prefer: return=representation
   * @see https://datatracker.ietf.org/doc/html/rfc7240
   */
  Prefer?: string;

  /**
   * Authorization credentials for connecting to a proxy.
   * @example Proxy-Authorization: Basic QWxhZGRpbjpvcGVuIHNlc2FtZQ==
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  "Proxy-Authorization"?: string;

  /**
   * Request only part of an entity. Bytes are numbered from 0. See Byte serving.
   * @example Range: bytes=500-999
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  Range?: string;

  /**
   * This is the address of the previous web page from which a link to the currently requested page was followed.
   * (The word "referrer" has been misspelled in the RFC as well as in most implementations to the point that
   * it has become standard usage and is considered correct terminology)
   * @example Referer: http://en.wikipedia.org/wiki/Main_Page
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  Referer?: string;

  /**
   * The transfer encodings the user agent is willing to accept:
   * the same values as for the response header field Transfer-Encoding can be used,
   * plus the "trailers" value (related to the "chunked" transfer method) to notify
   * the server it expects to receive additional fields in the trailer after the last, zero-sized, chunk.
   * Only trailers is supported in HTTP/2.
   * @example TE: trailers, deflate
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  TE?: string;

  /**
   * The Trailer general field value indicates that the given set of header fields is present
   * in the trailer of a message encoded with chunked transfer coding.
   * @example Trailer: Max-Forwards
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  Trailer?: string;

  /**
   * The form of encoding used to safely transfer the entity to the user.
   * Currently defined methods are: chunked, compress, deflate, gzip, identity.
   * Must not be used with HTTP/2.
   * @example Transfer-Encoding: chunked
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  "Transfer-Encoding"?: string;

  /**
   * The user agent string of the user agent.
   * @example User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:12.0) Gecko/20100101 Firefox/12.0
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  "User-Agent"?: string;

  /**
   * Ask the server to upgrade to another protocol. Must not be used in HTTP/2.
   * @example Upgrade: h2c, HTTPS/1.3, IRC/6.9, RTA/x11, websocket
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  Upgrade?: string;

  /**
   * Informs the server of proxies through which the request was sent.
   * @example Via: 1.0 fred, 1.1 example.com (Apache/1.1)
   * @see https://datatracker.ietf.org/doc/html/rfc9110
   */
  Via?: string;

  /**
   * A general warning about possible problems with the entity body.
   * @example Warning: 199 Miscellaneous warning
   * @see https://datatracker.ietf.org/doc/html/rfc7234, 9111
   */
  Warning?: string;

  /**
   * Mainly used to identify Ajax requests (most JavaScript frameworks send
   * this field with value of XMLHttpRequest); also identifies Android apps using WebView
   * @example X-Requested-With: XMLHttpRequest
   * @see https://datatracker.ietf.org/doc/html/rfc7234, 9111
   */
  "X-Requested-With"?: string;

  /**
   * Requests a web application to disable their tracking of a user.
   * This is Mozilla's version of the X-Do-Not-Track header field (since Firefox 4.0 Beta 11).
   * Safari and IE9 also have support for this field.
   * On March 7, 2011, a draft proposal was submitted to IETF.
   * The W3C Tracking Protection Working Group is producing a specification.
   * @example DNT: 1 (Do Not Track Enabled)
   * @see https://datatracker.ietf.org/doc/html/rfc7234, 9111
   */
  DNT?: string;

  /**
   * A de facto standard for identifying the originating IP address of a client connecting to a web server
   * through an HTTP proxy or load balancer. Superseded by Forwarded header.
   * @example X-Forwarded-For: client1, proxy1, proxy2
   * @see https://datatracker.ietf.org/doc/html/rfc7234, 9111
   */
  "X-Forwarded-For"?: string;

  /**
   * A de facto standard for identifying the original host requested by the client in
   * the Host HTTP request header, since the host name and/or port of the reverse proxy
   * (load balancer) may differ from the origin server handling the request.
   * Superseded by Forwarded header.
   * @example X-Forwarded-Host: en.wikipedia.org
   * @see https://datatracker.ietf.org/doc/html/rfc7234, 9111
   */
  "X-Forwarded-Host"?: string;

  /**
   * A de facto standard for identifying the originating protocol of an HTTP request,
   * since a reverse proxy (or a load balancer) may communicate with a web server using
   * HTTP even if the request to the reverse proxy is HTTPS. An alternative form of the header
   * (X-ProxyUser-Ip) is used by Google clients talking to Google servers.
   * Superseded by Forwarded header.
   * @example X-Forwarded-Proto: https
   * @see https://datatracker.ietf.org/doc/html/rfc7234, 9111
   */
  "X-Forwarded-Proto"?: string;

  /**
   * Requests a web application to override the method specified in the request (typically POST)
   * with the method given in the header field (typically PUT or DELETE).
   * This can be used when a user agent or firewall prevents PUT or DELETE methods from being sent directly
   * (note that this is either a bug in the software component, which ought to be fixed, or an
   * intentional configuration, in which case bypassing it may be the wrong thing to do).
   * @example X-HTTP-Method-Override: DELETE
   * @see https://datatracker.ietf.org/doc/html/rfc7234, 9111
   */
  "X-Http-Method-Override"?: string;

  /**
   * Allows easier parsing of the MakeModel/Firmware that is usually found in the User-Agent String of AT&T Devices
   * @example X-Att-Deviceid: GT-P7320/P7320XXLPG
   * @see https://datatracker.ietf.org/doc/html/rfc7234, 9111
   */
  "X-ATT-DeviceId"?: string;

  /**
   * Links to an XML file on the Internet with a full description and details about the
   * device currently connecting. In the example to the right is an XML file for an AT&T Samsung Galaxy S2.
   * @example x-wap-profile: http://wap.samsungmobile.com/uaprof/SGH-I777.xml
   * @see https://datatracker.ietf.org/doc/html/rfc7234, 9111
   */
  "X-Wap-Profile"?: string;

  /**
   * Implemented as a misunderstanding of the HTTP specifications. Common because of mistakes in
   * implementations of early HTTP versions. Has exactly the same functionality as standard Connection field.
   * Must not be used with HTTP/2.
   * @example Proxy-Connection: keep-alive
   * @see https://datatracker.ietf.org/doc/html/rfc7234, 9111
   */
  "Proxy-Connection"?: string;

  /**
   * Used to prevent cross-site request forgery. Alternative header names are: X-CSRFToken and X-XSRF-TOKEN
   * @example X-Csrf-Token: i8XNjC4b8KVok4uw5RftR38Wgp2BFwql
   * @see https://datatracker.ietf.org/doc/html/rfc7234, 9111
   */
  "X-Csrf-Token"?: string;

  /**
   * Correlates HTTP requests between a client and server.
   * @example X-Request-ID: f058ebd6-02f7-4d3f-942e-904344e8cde5
   * @see https://datatracker.ietf.org/doc/html/rfc7234, 9111
   */
  "X-Request-ID"?: string;

  /**
   * The Save-Data client hint request header available in Chrome, Opera, and Yandex browsers
   * lets developers deliver lighter, faster applications to users who opt-in to data saving mode in their browser.
   * @example Save-Data: on
   * @see https://datatracker.ietf.org/doc/html/rfc7234, 9111
   */
  "Save-Data"?: string;

  /**
   * Custom headers
   * @example X-My-Name: John Doe
   */
  [key: string]: string | undefined;
}

export default HttpRequestHeaders;
