// @triplyetl/vocabularies/src/Vocabularies/Generated/ceo.ts
/**
 * ceo
 * ceo.ts (2023-08-08T17:30:08.229Z)
 */
import VocabularyBase from '../../Classes/Vocabulary.js'

class Vocabulary extends VocabularyBase<'https://linkeddata.cultureelerfgoed.nl/def/ceo#'> {
  /**
   * **[ceo:Actor and role](https://linkeddata.cultureelerfgoed.nl/def/ceo#ActorEnRol)**
   *
   * The class ActorEnRol is the representation of the actor and role with respect to an event.
   *
   * @type        class
   * @parent      http://www.w3.org/2002/07/owl#Thing
   */
  public readonly ActorEnRol = this.toIri().concat('ActorEnRol')

  /**
   * **[ceo:Archaeological complex](https://linkeddata.cultureelerfgoed.nl/def/ceo#ArcheologischComplex)**
   *
   * The class ArcheologischComplex (ArchaeologicalComplex) is the representation of a collection, in time and space, of associated finds and/or archaeological soil traces.
   *
   * @type        class
   * @parent      ceo:CultuurhistorischObject
   */
  public readonly ArcheologischComplex = this.toIri().concat('ArcheologischComplex')

  /**
   * **[ceo:ArchaeologicalResearchArea](https://linkeddata.cultureelerfgoed.nl/def/ceo#ArcheologischOnderzoeksgebied)**
   *
   * The class ArcheologischOnderzoeksgebied (ArchaeologicalResearchArea) is the representation of a spatially bounded area, in which archaeological research is carried out by the permit holder to conduct research.
   *
   * @type        class
   * @parent      ceo:CultuurhistorischObject
   */
  public readonly ArcheologischOnderzoeksgebied = this.toIri().concat('ArcheologischOnderzoeksgebied')

  /**
   * **[ceo:Archeologisch terrein](https://linkeddata.cultureelerfgoed.nl/def/ceo#ArcheologischTerrein)**
   *
   * De class ArcheologischTerrein is de representatie van een ruimtelijk begrensd gebied waarbinnen zich archeologische informatie bevind.
   *
   * @type        class
   * @parent      ceo:CultuurhistorischObject
   */
  public readonly ArcheologischTerrein = this.toIri().concat('ArcheologischTerrein')

  /**
   * **[ceo:BAG relatie](https://linkeddata.cultureelerfgoed.nl/def/ceo#BAGRelatie)**
   *
   * De class BAGRelatie is de representatie van vastgelegde Basisregistratie Adressen en Gebouwen (BAG) met betrekking tot een cultuurhistorisch object.
   *
   * @type        class
   * @parent      http://www.w3.org/2002/07/owl#Thing
   */
  public readonly BAGRelatie = this.toIri().concat('BAGRelatie')

  /**
   * **[ceo:BAG relatienummer](https://linkeddata.cultureelerfgoed.nl/def/ceo#BAGRelatienummer)**
   *
   * De property BAGRelatienummer wordt gebruikt om aan te geven het unieke nummer waaronder de BAG relatie instance bekend is.
   *
   * @type        property
   * @domain      ceo:BAGRelatie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly BAGRelatienummer = this.toIri().concat('BAGRelatienummer')

  /**
   * **[ceo:BRK relatie](https://linkeddata.cultureelerfgoed.nl/def/ceo#BRKRelatie)**
   *
   * De class BRKRelatie is de representatie van vastgelegde Basisregistratie Kadaster (BRK) met betrekking tot een cultuurhistorisch object.
   *
   * @type        class
   * @parent      http://www.w3.org/2002/07/owl#Thing
   */
  public readonly BRKRelatie = this.toIri().concat('BRKRelatie')

  /**
   * **[ceo:BRK relatienummer](https://linkeddata.cultureelerfgoed.nl/def/ceo#BRKRelatienummer)**
   *
   * De property BRKRelatienummer wordt gebruikt om aan te geven het unieke nummer waaronder de BRK relatie instance bekend is.
   *
   * @type        property
   * @domain      ceo:BRKRelatie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly BRKRelatienummer = this.toIri().concat('BRKRelatienummer')

  /**
   * **[ceo:Basisregistratie relatie](https://linkeddata.cultureelerfgoed.nl/def/ceo#BasisregistratieRelatie)**
   *
   * De class BasisregistratieRelatie is de representatie van de administratieve vastlegging van de relatie tussen een cultuurhistorisch object en een instantie in een basisregistratie.
   *
   * @type        class
   * @parent      http://www.w3.org/2002/07/owl#Thing
   */
  public readonly BasisregistratieRelatie = this.toIri().concat('BasisregistratieRelatie')

  /**
   * **[ceo:Begindatering](https://linkeddata.cultureelerfgoed.nl/def/ceo#BeginDatering)**
   *
   * De class BeginDatering is de representatie van de begin datering van de tijdsaanduiding die een Cultuurhistorisch Object kan hebben.
   *
   * @type        class
   */
  public readonly BeginDatering = this.toIri().concat('BeginDatering')

  /**
   * **[ceo:Bouwkundige kwaliteit](https://linkeddata.cultureelerfgoed.nl/def/ceo#BouwkundigeKwaliteit)**
   *
   * De class BouwkundigeKwaliteit is een subclass van Kennisregistratie en is de representatie van de bouwkundige kwaliteit met betrekking tot een cultuurhistorisch object.
   *
   * @type        class
   * @parent      ceo:Kennisregistratie
   */
  public readonly BouwkundigeKwaliteit = this.toIri().concat('BouwkundigeKwaliteit')

  /**
   * **[ceo:Bronverwijzing](https://linkeddata.cultureelerfgoed.nl/def/ceo#Bronverwijzing)**
   *
   * De class Bronverwijzing is een subclass van Kennisregistratie en is de representatie van de bronverwijzing met betrekking tot een cultuurhistorisch object.
   *
   * @type        class
   * @parent      http://www.w3.org/2002/07/owl#Thing
   */
  public readonly Bronverwijzing = this.toIri().concat('Bronverwijzing')

  /**
   * **[ceo:Complex](https://linkeddata.cultureelerfgoed.nl/def/ceo#Complex)**
   *
   * De class Complex is de representatie van beschrijvende of waarderende samenhang tussen rijksmonumenten, die als een complex zijn aangewezen. Deze samenhang is niet noodzakelijk voor de identificatie van die rijksmonumenten.
   *
   * @type        class
   * @parent      ceo:CultuurhistorischObject
   */
  public readonly Complex = this.toIri().concat('Complex')

  /**
   * **[ceo:Creative Common License 4.0](https://linkeddata.cultureelerfgoed.nl/def/ceo#'CreativeCommonsLicense4.0')**
   *
   * @type        instance
   */
  public readonly 'CreativeCommonsLicense4.0' = this.toIri().concat('CreativeCommonsLicense4.0')

  /**
   * **[ceo:Culturalhistorical object](https://linkeddata.cultureelerfgoed.nl/def/ceo#CultuurhistorischObject)**
   *
   * The class CultuurhistorischObject (Cultural-historical Object) is the representation of a phenomenon in reality with a certain cultural-historical value.
   *
   * @type        class
   * @parent      http://www.w3.org/2002/07/owl#Thing
   */
  public readonly CultuurhistorischObject = this.toIri().concat('CultuurhistorischObject')

  /**
   * **[ceo:Datering](https://linkeddata.cultureelerfgoed.nl/def/ceo#Datering)**
   *
   * De class Datering is een representatie van de tijdsaanduiding die een Cultuurhistorisch Object kan hebben.
   *
   * @type        class
   * @parent      http://www.w3.org/2002/07/owl#Thing
   */
  public readonly Datering = this.toIri().concat('Datering')

  /**
   * **[ceo:Einddatering](https://linkeddata.cultureelerfgoed.nl/def/ceo#EindDatering)**
   *
   * De class EindDatering is de representatie van de eind datering van de tijdsaanduiding die een Cultuurhistorisch Object kan hebben.
   *
   * @type        class
   */
  public readonly EindDatering = this.toIri().concat('EindDatering')

  /**
   * **[ceo:Functie](https://linkeddata.cultureelerfgoed.nl/def/ceo#Functie)**
   *
   * De class Functies is een subclass van kennisregistratie en is de representatie van de functie met betrekking tot een cultuurhistorisch object.
   *
   * @type        class
   * @parent      ceo:Kennisregistratie
   */
  public readonly Functie = this.toIri().concat('Functie')

  /**
   * **[ceo:Gebeurtenis](https://linkeddata.cultureelerfgoed.nl/def/ceo#Gebeurtenis)**
   *
   * De class Gebeurtenis is een subclass van kennisregistratie en is de representatie van vastgelegde gebeurtenissen met betrekking tot een cultuurhistorisch object.
   *
   * @type        class
   * @parent      ceo:Kennisregistratie
   */
  public readonly Gebeurtenis = this.toIri().concat('Gebeurtenis')

  /**
   * **[ceo:Geometrie](https://linkeddata.cultureelerfgoed.nl/def/ceo#Geometrie)**
   *
   * De class Geometrie is de representatie van de geometrie van een cultuurhistorisch object.
   *
   * @type        class
   * @parent      http://www.opengis.net/ont/geosparql#Geometry
   */
  public readonly Geometrie = this.toIri().concat('Geometrie')

  /**
   * **[ceo:Gezicht](https://linkeddata.cultureelerfgoed.nl/def/ceo#Gezicht)**
   *
   * De class Gezicht is een representatie van een aangewezen, beschermd stads- en dorpsgezicht.
   *
   * @type        class
   * @parent      ceo:CultuurhistorischObject
   */
  public readonly Gezicht = this.toIri().concat('Gezicht')

  /**
   * **[ceo:Grondsporen](https://linkeddata.cultureelerfgoed.nl/def/ceo#Grondsporen)**
   *
   * De class Grondsporen is een representatie van een duidelijk begrensbaar verschijnsel van antropogene of natuurlijke oorsprong.
   *
   * @type        class
   * @parent      ceo:CultuurhistorischObject
   */
  public readonly Grondsporen = this.toIri().concat('Grondsporen')

  /**
   * **[ceo:Kennisregistratie](https://linkeddata.cultureelerfgoed.nl/def/ceo#Kennisregistratie)**
   *
   * De class Kennisregistratie is de representatie van vastgelegde kennis feiten met betrekking tot een cultuurhistorisch object.
   *
   * @type        class
   */
  public readonly Kennisregistratie = this.toIri().concat('Kennisregistratie')

  /**
   * **[ceo:Locatie aanduiding](https://linkeddata.cultureelerfgoed.nl/def/ceo#LocatieAanduiding)**
   *
   * De class LocatieAanduiding is de representatie van de locatie aanduiding met betrekking tot een cultuurhistorisch object.
   *
   * @type        class
   * @parent      http://www.w3.org/2002/07/owl#Thing
   */
  public readonly LocatieAanduiding = this.toIri().concat('LocatieAanduiding')

  /**
   * **[ceo:Materiaal](https://linkeddata.cultureelerfgoed.nl/def/ceo#Materiaal)**
   *
   * De class Materiaal is een subclass van kennisregistratie en is de representatie van vastgelegde materialen met betrekking tot een cultuurhistorisch object.
   *
   * @type        class
   * @parent      ceo:Kennisregistratie
   */
  public readonly Materiaal = this.toIri().concat('Materiaal')

  /**
   * **[ceo:Naam](https://linkeddata.cultureelerfgoed.nl/def/ceo#Naam)**
   *
   * De class Naam is een subclass van kennisregistratie en is de representatie van vastgelegde namen met betrekking tot een cultuurhistorisch object.
   *
   * @type        class
   * @parent      ceo:Kennisregistratie
   */
  public readonly Naam = this.toIri().concat('Naam')

  /**
   * **[ceo:Omschrijving](https://linkeddata.cultureelerfgoed.nl/def/ceo#Omschrijving)**
   *
   * De class Omschrijving is een subclass van kennisregistratie en is de representatie van vastgelegde omschrijvingen met betrekking tot een cultuurhistorisch object.
   *
   * @type        class
   * @parent      ceo:Kennisregistratie
   */
  public readonly Omschrijving = this.toIri().concat('Omschrijving')

  /**
   * **[ceo:Rijksdienst_voor_het_Cultureel_Erfgoed](https://linkeddata.cultureelerfgoed.nl/def/ceo#Rijksdienst_voor_het_Cultureel_Erfgoed)**
   *
   * @type        instance
   */
  public readonly Rijksdienst_voor_het_Cultureel_Erfgoed = this.toIri().concat('Rijksdienst_voor_het_Cultureel_Erfgoed')

  /**
   * **[ceo:Rijksmonument](https://linkeddata.cultureelerfgoed.nl/def/ceo#Rijksmonument)**
   *
   * De class Rijksmonument is een representatie van onroerende cultuurhistorische objecten die als rijksmonument zijn aangewezen.
   *
   * @type        class
   * @parent      ceo:CultuurhistorischObject
   */
  public readonly Rijksmonument = this.toIri().concat('Rijksmonument')

  /**
   * **[ceo:Stijl en cultuur](https://linkeddata.cultureelerfgoed.nl/def/ceo#StijlEnCultuur)**
   *
   * De class StijlEnCultuur is een subclass van kennisregistratie en is de representatie van de stijl en cultuur met betrekking tot een cultuurhistorisch object.
   *
   * @type        class
   * @parent      ceo:Kennisregistratie
   */
  public readonly StijlEnCultuur = this.toIri().concat('StijlEnCultuur')

  /**
   * **[ceo:Type](https://linkeddata.cultureelerfgoed.nl/def/ceo#Type)**
   *
   * De class Type is een subclass van kennisregistratie en is de representatie van vastgelegde typen met betrekking tot een cultuurhistorisch object.
   *
   * @type        class
   * @parent      ceo:Kennisregistratie
   */
  public readonly Type = this.toIri().concat('Type')

  /**
   * **[ceo:Vondsten](https://linkeddata.cultureelerfgoed.nl/def/ceo#Vondsten)**
   *
   * De class Vondsten is een representatie van alle soorten mobilia: roerende of roerend geraakte onderdelen van onroerende goederen, afkomstig van archeologisch veldwerk of archeologische collecties.
   *
   * @type        class
   * @parent      ceo:CultuurhistorischObject
   */
  public readonly Vondsten = this.toIri().concat('Vondsten')

  /**
   * **[ceo:Vondstlocatie](https://linkeddata.cultureelerfgoed.nl/def/ceo#Vondstlocatie)**
   *
   * De class Vondstlocatie is een representatie van geografisch begrensde gebieden, waar binnen min of meer aaneensluitend archeologisch sporen en/of vondsten zijn aangetroffen, eventueel van verschillende ouderdom.
   *
   * @type        class
   * @parent      ceo:CultuurhistorischObject
   */
  public readonly Vondstlocatie = this.toIri().concat('Vondstlocatie')

  /**
   * **[ceo:Werelderfgoed](https://linkeddata.cultureelerfgoed.nl/def/ceo#Werelderfgoed)**
   *
   * De class Werelderfgoed is een representatie van een door het Werelderfgoedcomité van UNESCO op de Werelderfgoedlijst geplaats cultuurhistorisch object.
   *
   * @type        class
   * @parent      ceo:CultuurhistorischObject
   */
  public readonly Werelderfgoed = this.toIri().concat('Werelderfgoed')

  /**
   * **[ceo:Aantal grondsporen](https://linkeddata.cultureelerfgoed.nl/def/ceo#aantalGrondsporen)**
   *
   * De property aantalGrondsporen wordt gebruikt om aan te geven het aantal grondsporen.
   *
   * @type        property
   * @domain      ceo:Grondsporen
   * @range       http://www.w3.org/2001/XMLSchema#integer
   */
  public readonly aantalGrondsporen = this.toIri().concat('aantalGrondsporen')

  /**
   * **[ceo:Aantal vondsten](https://linkeddata.cultureelerfgoed.nl/def/ceo#aantalVondsten)**
   *
   * De property aantalVondsten wordt gebruikt om aan te geven het aantal vondsten.
   *
   * @type        property
   * @domain      ceo:Vondsten
   * @range       http://www.w3.org/2001/XMLSchema#integer
   */
  public readonly aantalVondsten = this.toIri().concat('aantalVondsten')

  /**
   * **[ceo:Actor en rolnummer](https://linkeddata.cultureelerfgoed.nl/def/ceo#actorEnRolnummer)**
   *
   * De property actorEnRolnummer wordt gebruikt om aan te geven het unieke nummer waaronder de actor en rol instance bekend is.
   *
   * @type        property
   * @domain      ceo:ActorEnRol
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly actorEnRolnummer = this.toIri().concat('actorEnRolnummer')

  /**
   * **[ceo:Appartement aanduiding](https://linkeddata.cultureelerfgoed.nl/def/ceo#appartementAanduiding)**
   *
   * De property appartementAanduiding wordt gebruikt om aan te geven of een appartement aanduiding aan een rijksmonument is gegeven.
   *
   * @type        property
   * @domain      ceo:BRKRelatie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly appartementAanduiding = this.toIri().concat('appartementAanduiding')

  /**
   * **[ceo:Archis2 complexnummer](https://linkeddata.cultureelerfgoed.nl/def/ceo#archis2Complexnummer)**
   *
   * De property archis2Complexnummer wordt gebruikt om aan te geven of een complexnummer uit Archis 2 aan een archeologisch complex of vondstlocatie is gekoppeld.
   *
   * @type        property
   * @domain      ceo:ArcheologischComplex or ceo:Vondstlocatie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly archis2Complexnummer = this.toIri().concat('archis2Complexnummer')

  /**
   * **[ceo:Archis2 monumentnummer](https://linkeddata.cultureelerfgoed.nl/def/ceo#archis2Monumentnummer)**
   *
   * De property archis2Monumentnummer wordt gebruikt om aan te geven of een monumentnummer uit Archis 2 aan een archeologisch terrein is gekoppeld.
   *
   * @type        property
   * @domain      ceo:ArcheologischTerrein
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly archis2Monumentnummer = this.toIri().concat('archis2Monumentnummer')

  /**
   * **[ceo:Archis2 vondstmeldingsnummer](https://linkeddata.cultureelerfgoed.nl/def/ceo#archis2Vondstmeldingsnummer)**
   *
   * De property archis2Vondstmeldingsnummer wordt gebruikt om aan te geven of een vondstmeldingsnummer uit Archis 2 aan een vondstlocatie is gekoppeld.
   *
   * @type        property
   * @domain      ceo:Vondstlocatie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly archis2Vondstmeldingsnummer = this.toIri().concat('archis2Vondstmeldingsnummer')

  /**
   * **[ceo:Archis2 vondstnummer](https://linkeddata.cultureelerfgoed.nl/def/ceo#archis2Vondstnummer)**
   *
   * De property archis2Vondstnummer wordt gebruikt om aan te geven of een vondstnummer uit Archis 2 aan een vondst, grondspoor is gekoppeld.
   *
   * @type        property
   * @domain      ceo:Grondsporen or ceo:Vondsten
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly archis2Vondstnummer = this.toIri().concat('archis2Vondstnummer')

  /**
   * **[ceo:Archis2 waarnemingsnummer](https://linkeddata.cultureelerfgoed.nl/def/ceo#archis2Waarnemingsnummer)**
   *
   * De property archis2Waarnemingsnummer wordt gebruikt om aan te geven of een waarnemingsnummer uit Archis 2 aan een vondstlocatie is gekoppeld.
   *
   * @type        property
   * @domain      ceo:Vondstlocatie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly archis2Waarnemingsnummer = this.toIri().concat('archis2Waarnemingsnummer')

  /**
   * **[ceo:asWKT-RD](https://linkeddata.cultureelerfgoed.nl/def/ceo#'asWKT-RD')**
   *
   * De geometrie is een registergegeven binnen de class Rijksmonument.
   *
   * @type        property
   * @domain      ceo:Geometrie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly 'asWKT-RD' = this.toIri().concat('asWKT-RD')

  /**
   * **[ceo:Basisregistratie relatienummer](https://linkeddata.cultureelerfgoed.nl/def/ceo#basisregistratieRelatienummer)**
   *
   * De property basisregistratieRelatienummer wordt gebruikt om aan te geven het unieke nummer waaronder de basisregistratie relatie instance bekend is.
   *
   * @type        property
   * @domain      ceo:BasisregistratieRelatie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly basisregistratieRelatienummer = this.toIri().concat('basisregistratieRelatienummer')

  /**
   * **[ceo:Bevat object](https://linkeddata.cultureelerfgoed.nl/def/ceo#bevatObject)**
   *
   * De property bevatObject legt de relatie tussen de ligging van een cultuurhistorisch object met betrekking tot een cultuurhistorisch object vast.
   *
   * @type        property
   * @domain      ceo:CultuurhistorischObject
   * @range       ceo:CultuurhistorischObject
   */
  public readonly bevatObject = this.toIri().concat('bevatObject')

  /**
   * **[ceo:Bijzonder gebied](https://linkeddata.cultureelerfgoed.nl/def/ceo#bijzonderGebied)**
   *
   * e property bijzonderGebied wordt gebruikt om aan te geven dat een beschermd stads- of dorpsgezicht een gebied in een stad of dorp is met een bijzonder cultuurhistorisch karakter.
   *
   * @type        property
   * @domain      ceo:Gezicht
   * @range       http://www.w3.org/2001/XMLSchema#boolean
   */
  public readonly bijzonderGebied = this.toIri().concat('bijzonderGebied')

  /**
   * **[ceo:Complexnummer](https://linkeddata.cultureelerfgoed.nl/def/ceo#complexnummer)**
   *
   * De property complexnummer wordt gebruikt om aan te geven het unieke nummer waaronder het complex van rijksmonumenten bekend is in het monumentenregister.
   *
   * @type        property
   * @domain      ceo:Complex
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly complexnummer = this.toIri().concat('complexnummer')

  /**
   * **[ceo:Cultuurhistorisch objectnummer](https://linkeddata.cultureelerfgoed.nl/def/ceo#cultuurhistorischObjectnummer)**
   *
   * De property cultuurhistorischObjectnummer wordt gebruikt om aan te geven of een cultuurhistorischObjectnummer aan een Cultuurhistorisch object is gegeven.
   *
   * @type        property
   * @domain      ceo:CultuurhistorischObject
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly cultuurhistorischObjectnummer = this.toIri().concat('cultuurhistorischObjectnummer')

  /**
   * **[ceo:Dateringnummer](https://linkeddata.cultureelerfgoed.nl/def/ceo#dateringnummer)**
   *
   * De property dateringnummer wordt gebruikt om aan te geven het unieke nummer waaronder de datering instance bekend is.
   *
   * @type        property
   * @domain      ceo:Datering
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly dateringnummer = this.toIri().concat('dateringnummer')

  /**
   * **[ceo:datum](https://linkeddata.cultureelerfgoed.nl/def/ceo#datum)**
   *
   * De property datum wordt gebruikt om aan te geven de datum van een begin- of een einddatering.
   *
   * @type        property
   * @domain      ceo:BeginDatering or ceo:EindDatering
   * @range       http://www.w3.org/2001/XMLSchema#date
   */
  public readonly datum = this.toIri().concat('datum')

  /**
   * **[ceo:Datum inschrijving in Monumentenregister](https://linkeddata.cultureelerfgoed.nl/def/ceo#datumInschrijvingInMonumentenregister)**
   *
   * De datum van inschrijving in monumentenregister is een registergegeven binnen de class Rijksmonument.
   *
   * @type        property
   * @domain      ceo:Rijksmonument
   * @range       http://www.w3.org/2001/XMLSchema#date
   */
  public readonly datumInschrijvingInMonumentenregister = this.toIri().concat('datumInschrijvingInMonumentenregister')

  /**
   * **[ceo:Formeel standpunt](https://linkeddata.cultureelerfgoed.nl/def/ceo#formeelStandpunt)**
   *
   * De property formeelstandpunt wordt gebruikt om aan te geven of de kennisregistratie een formeel standpunt betreft.
   *
   * @type        property
   * @domain      ceo:Kennisregistratie
   * @range       http://www.w3.org/2001/XMLSchema#boolean
   */
  public readonly formeelStandpunt = this.toIri().concat('formeelStandpunt')

  /**
   * **[ceo:Gemeentecode](https://linkeddata.cultureelerfgoed.nl/def/ceo#gemeentecode)**
   *
   * De property BRK gemeenteCode wordt gebruikt om aan te geven welke gemeente code is gekoppeld aan het rijksmonument.
   *
   * @type        property
   * @domain      ceo:BRKRelatie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly gemeentecode = this.toIri().concat('gemeentecode')

  /**
   * **[ceo:Gemeentenaam](https://linkeddata.cultureelerfgoed.nl/def/ceo#gemeentenaam)**
   *
   * De property BRK gemeentenaam wordt gebruikt om aan te geven welke gemeentenaam is gekoppeld aan het rijksmonument.
   *
   * @type        property
   * @domain      ceo:BRKRelatie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly gemeentenaam = this.toIri().concat('gemeentenaam')

  /**
   * **[ceo:Geometriebesluit](https://linkeddata.cultureelerfgoed.nl/def/ceo#geometriebesluit)**
   *
   * De property geometriebesluit wordt gebruikt om aan te geven de besluit status van de geometrie.
   *
   * @type        property
   * @domain      ceo:Geometrie
   * @range       http://www.w3.org/2001/XMLSchema#boolean
   */
  public readonly geometriebesluit = this.toIri().concat('geometriebesluit')

  /**
   * **[ceo:Geometrienummer](https://linkeddata.cultureelerfgoed.nl/def/ceo#geometrienummer)**
   *
   * De property geometrienummer wordt gebruikt om aan te geven het unieke nummer waaronder de geometrie instance bekend is.
   *
   * @type        property
   * @domain      ceo:Geometrie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly geometrienummer = this.toIri().concat('geometrienummer')

  /**
   * **[ceo:Gezichtsnummer](https://linkeddata.cultureelerfgoed.nl/def/ceo#gezichtsnummer)**
   *
   * De property gezichtsnummer wordt gebruikt om aan te geven het unieke nummer waaronder het stads- of dorpsgezicht bekend is.
   *
   * @type        property
   * @domain      ceo:Gezicht
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly gezichtsnummer = this.toIri().concat('gezichtsnummer')

  /**
   * **[ceo:Heeft actor](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftActor)**
   *
   * De property heeftActor legt de relatie tussen de actor en rol en de actor met betrekking tot een cultuurhistorisch object vast.
   *
   * @type        property
   * @domain      ceo:ActorEnRol
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftActor = this.toIri().concat('heeftActor')

  /**
   * **[ceo:Heeft actor en rol](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftActorEnRol)**
   *
   * De property heeftActorEnRol legt de relatie tussen de gebeurtenis en de actor en zijn rol met betrekking tot een cultuurhistorisch object vast.
   *
   * @type        property
   * @domain      ceo:Gebeurtenis
   * @range       ceo:ActorEnRol
   */
  public readonly heeftActorEnRol = this.toIri().concat('heeftActorEnRol')

  /**
   * **[ceo:Heeft archeologische waardering](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftArcheologischeWaardering)**
   *
   * De property heeftArcheologischeWaardering legt de relatie tussen het archeologisch terrein en de archeologische waardering vast.
   *
   * @type        property
   * @domain      ceo:ArcheologischTerrein
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftArcheologischeWaardering = this.toIri().concat('heeftArcheologischeWaardering')

  /**
   * **[ceo:Heeft BAG relatie](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftBAGRelatie)**
   *
   * De property heeftBAGRelatie legt de relatie tussen de basisregistratie en de BAG relatie vast.
   *
   * @type        property
   * @domain      ceo:BasisregistratieRelatie
   * @range       ceo:BAGRelatie
   */
  public readonly heeftBAGRelatie = this.toIri().concat('heeftBAGRelatie')

  /**
   * **[ceo:Heeft BRK relatie](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftBRKRelatie)**
   *
   * De property heeftBRKRelatie legt de relatie tussen de basisregistratie en de BRK relatie vast.
   *
   * @type        property
   * @domain      ceo:BasisregistratieRelatie
   * @range       ceo:BRKRelatie
   */
  public readonly heeftBRKRelatie = this.toIri().concat('heeftBRKRelatie')

  /**
   * **[ceo:Heeft basisregistratie relatie](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftBasisregistratieRelatie)**
   *
   * De property heeftBasisregistratieRelatie legt de relatie tussen het cultuurhistorisch object en de basisregistratie relatie vast.
   *
   * @type        property
   * @domain      ceo:CultuurhistorischObject
   * @range       ceo:BasisregistratieRelatie
   */
  public readonly heeftBasisregistratieRelatie = this.toIri().concat('heeftBasisregistratieRelatie')

  /**
   * **[ceo:Heeft bebouwde kom type](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftBebouwdeKomType)**
   *
   * De property heeftBebouwdeKomtype legt de relatie tussen het rijksmonument en het bebouwde kom type vast.
   *
   * @type        property
   * @domain      ceo:Rijksmonument
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftBebouwdeKomType = this.toIri().concat('heeftBebouwdeKomType')

  /**
   * **[ceo:Heeft begindatering](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftBeginDatering)**
   *
   * De property heeftBeginDatering legt de relatie tussen de datering en de begindatering vast.
   *
   * @type        property
   * @domain      ceo:Datering
   * @range       ceo:BeginDatering
   */
  public readonly heeftBeginDatering = this.toIri().concat('heeftBeginDatering')

  /**
   * **[ceo:Heeft betrekking op](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftBetrekkingOp)**
   *
   * De property heeftBetrekkingOp legt de relatie tussen de kennisregistratie en het cultuurhistorisch object vast.
   *
   * @type        property
   * @domain      ceo:Kennisregistratie
   * @range       ceo:CultuurhistorischObject
   */
  public readonly heeftBetrekkingOp = this.toIri().concat('heeftBetrekkingOp')

  /**
   * **[ceo:Heeft betrouwbaarheid](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftBetrouwbaarheid)**
   *
   * De property heeftBetrouwbaarheid legt de relatie tussen de datering en de mate van betrouwbaarheid vast.
   *
   * @type        property
   * @domain      ceo:Datering
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftBetrouwbaarheid = this.toIri().concat('heeftBetrouwbaarheid')

  /**
   * **[ceo:Heeft bouwkundige kwaliteit](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftBouwkundigeKwaliteit)**
   *
   * De property heeftBouwkundigeKwaliteit legt de relatie tussen een cultuurhistorisch object en de bouwkundige kwaliteit vast.
   *
   * @type        property
   * @domain      ceo:CultuurhistorischObject
   * @range       ceo:BouwkundigeKwaliteit
   */
  public readonly heeftBouwkundigeKwaliteit = this.toIri().concat('heeftBouwkundigeKwaliteit')

  /**
   * **[ceo:Heeft bouwkundige staat](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftBouwkundigeStaat)**
   *
   * De property heeftBouwkundigeStaat legt de relatie tussen de BouwkundigeKwaliteit en de bouwkundige staat met betrekking tot een cultuurhistorisch object vast.
   *
   * @type        property
   * @domain      ceo:BouwkundigeKwaliteit
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftBouwkundigeStaat = this.toIri().concat('heeftBouwkundigeStaat')

  /**
   * **[ceo:Heeft datering](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftDatering)**
   *
   * De property heeftDatering legt de relatie tussen een gebeurtenis en de datering met betrekking tot een cultuurhistorisch object vast.
   *
   * @type        property
   * @domain      ceo:Gebeurtenis
   * @range       ceo:Datering
   */
  public readonly heeftDatering = this.toIri().concat('heeftDatering')

  /**
   * **[ceo:Heeft dateringsmethode](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftDateringsmethode)**
   *
   * De property heeftDatering legt de relatie tussen een gebeurtenis en de dateringsmethode vast.
   *
   * @type        property
   * @domain      ceo:Datering
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftDateringsmethode = this.toIri().concat('heeftDateringsmethode')

  /**
   * **[ceo:Heeft einddatering](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftEindDatering)**
   *
   * De property heeftBeginDatering legt de relatie tussen de datering en de einddatering vast.
   *
   * @type        property
   * @domain      ceo:Datering
   * @range       ceo:EindDatering
   */
  public readonly heeftEindDatering = this.toIri().concat('heeftEindDatering')

  /**
   * **[ceo:Heeft functie](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftFunctie)**
   *
   * De property heeftFunctie legt de relatie tussen een cultuurhistorisch object en de functie vast.
   *
   * @type        property
   * @domain      ceo:CultuurhistorischObject
   * @range       ceo:Functie
   */
  public readonly heeftFunctie = this.toIri().concat('heeftFunctie')

  /**
   * **[ceo:Heeft functienaam](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftFunctieNaam)**
   *
   * De property heeftFunctienaam legt de relatie tussen de functie en de naam van de functie vast.
   *
   * @type        property
   * @domain      ceo:Functie
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftFunctieNaam = this.toIri().concat('heeftFunctieNaam')

  /**
   * **[ceo:Heeft gebeurtenis](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftGebeurtenis)**
   *
   * De property heeftGebeurtenis legt de relatie tussen een cultuurhistorisch object en de gebeurtenis vast.
   *
   * @type        property
   * @domain      ceo:CultuurhistorischObject
   * @range       ceo:Gebeurtenis
   */
  public readonly heeftGebeurtenis = this.toIri().concat('heeftGebeurtenis')

  /**
   * **[ceo:Heeft gebeurtenisnaam](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftGebeurtenisNaam)**
   *
   * De property heeftGebeurtenisNaam legt de relatie tussen de gebeurtenis en de naam van de gebeurtenis vast.
   *
   * @type        property
   * @domain      ceo:Gebeurtenis
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftGebeurtenisNaam = this.toIri().concat('heeftGebeurtenisNaam')

  /**
   * **[ceo:Heeft gemeente](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftGemeente)**
   *
   * De property heeftGemeente is een registergegeven binnen de class Rijksmonument.
   *
   * @type        property
   * @domain      ceo:BasisregistratieRelatie
   * @range       http://standaarden.overheid.nl/owms/terms/Gemeente
   */
  public readonly heeftGemeente = this.toIri().concat('heeftGemeente')

  /**
   * **[ceo:Heeft geometrie](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftGeometrie)**
   *
   * De property heeftGeometrie legt de relatie tussen een cultuurhistorisch object en de geometrie vast.
   *
   * @type        property
   * @parent      http://www.opengis.net/ont/geosparql#hasGeometry
   * @domain      ceo:CultuurhistorischObject
   * @range       ceo:Geometrie
   */
  public readonly heeftGeometrie = this.toIri().concat('heeftGeometrie')

  /**
   * **[ceo:Heeft geometrie kwaliteit](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftGeometrieKwaliteit)**
   *
   * De property heeftGeometrieKwaliteit legt de relatie tussen de geometrie en de geometrische kwaliteit vast.
   *
   * @type        property
   * @domain      ceo:Geometrie
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftGeometrieKwaliteit = this.toIri().concat('heeftGeometrieKwaliteit')

  /**
   * **[ceo:Heeft geometrie status](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftGeometrieStatus)**
   *
   * De property heeftGeometrieStatus legt de relatie tussen de geometrie en de geometrische status vast.
   *
   * @type        property
   * @domain      ceo:Geometrie
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftGeometrieStatus = this.toIri().concat('heeftGeometrieStatus')

  /**
   * **[ceo:Heeft gezichtsstatus](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftGezichtsstatus)**
   *
   * De property heeftGezichtsstatus legt de relatie tussen het gezicht en het gezichtsstatus vast.
   *
   * @type        property
   * @domain      ceo:Gezicht
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftGezichtsstatus = this.toIri().concat('heeftGezichtsstatus')

  /**
   * **[ceo:Heeft herkomst](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftHerkomst)**
   *
   * De property heeftHerkomst legt de relatie tussen de geometrie en de herkomst vast.
   *
   * @type        property
   * @domain      ceo:Geometrie
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftHerkomst = this.toIri().concat('heeftHerkomst')

  /**
   * **[ceo:Heeft hoofdobject](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftHoofdobject)**
   *
   * De property heeftHoofdobject legt de relatie tussen het complex en het hoofdobject met betrekking tot een rijksmonument vast.
   *
   * @type        property
   * @domain      ceo:Complex
   * @range       ceo:Rijksmonument
   */
  public readonly heeftHoofdobject = this.toIri().concat('heeftHoofdobject')

  /**
   * **[ceo:Heeft huidige functie](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftHuidigeFunctie)**
   *
   * De property heeftHuidigeFunctie legt de relatie tussen een cultuurhistorisch object en de huidige functie vast.
   *
   * @type        property
   * @parent      ceo:heeftFunctie
   * @domain      ceo:Rijksmonument
   * @range       ceo:Functie
   */
  public readonly heeftHuidigeFunctie = this.toIri().concat('heeftHuidigeFunctie')

  /**
   * **[ceo:Heeft indicatie nauwkeurigheid](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftIndicatieNauwkeurigheid)**
   *
   * De property heeftIndicatieNauwkeurigheid legt de relatie tussen de datering en de indicatie nauwkeurigheid vast.
   *
   * @type        property
   * @domain      ceo:Datering
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftIndicatieNauwkeurigheid = this.toIri().concat('heeftIndicatieNauwkeurigheid')

  /**
   * **[ceo:Heeft juridische status](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftJuridischeStatus)**
   *
   * De property heeftJuridischeStatus legt de relatie het rijksmonument en de juridische status, geeft aan of het rijksmonument voorbeschermd of rijksmonument is.
   *
   * @type        property
   * @domain      ceo:Rijksmonument
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftJuridischeStatus = this.toIri().concat('heeftJuridischeStatus')

  /**
   * **[ceo:Heeft kennisdomein](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftKennisdomein)**
   *
   * De property heeftKennisdomein legt de relatie tussen de kennisregistratie en het kennisdomein vast.
   *
   * @type        property
   * @domain      ceo:Kennisregistratie
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftKennisdomein = this.toIri().concat('heeftKennisdomein')

  /**
   * **[ceo:Heeft kennisregistratie](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftKennisregistratie)**
   *
   * De property heeftKennisregistratie legt de relatie tussen het cultuurhistorisch object en de Kennisregistratie vast.
   *
   * @type        property
   * @domain      ceo:CultuurhistorischObject
   * @range       ceo:Kennisregistratie
   */
  public readonly heeftKennisregistratie = this.toIri().concat('heeftKennisregistratie')

  /**
   * **[ceo:Heeft locatie aanduiding](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftLocatieAanduiding)**
   *
   * De property heeftLocatieAanduiding legt de relatie tussen het cultuurhistorisch object en de locatie aanduiding vast.
   *
   * @type        property
   * @domain      ceo:CultuurhistorischObject
   * @range       ceo:LocatieAanduiding
   */
  public readonly heeftLocatieAanduiding = this.toIri().concat('heeftLocatieAanduiding')

  /**
   * **[ceo:Heeft locatie aanduiding status](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftLocatieAanduidingStatus)**
   *
   * De property heeftLocatieAanduidingStatus legt de relatie tussen de locatie aanduiding en de status van de locatie Aanduiding met betrekking tot een cultuurhistorisch object vast.
   *
   * @type        property
   * @domain      ceo:LocatieAanduiding
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftLocatieAanduidingStatus = this.toIri().concat('heeftLocatieAanduidingStatus')

  /**
   * **[ceo:Heeft materiaal](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftMateriaal)**
   *
   * De property heeftMateriaal legt de relatie tussen een cultuurhistorisch object en het materiaal vast.
   *
   * @type        property
   * @domain      ceo:CultuurhistorischObject
   * @range       ceo:Materiaal
   */
  public readonly heeftMateriaal = this.toIri().concat('heeftMateriaal')

  /**
   * **[ceo:Heeft materiaalnaam](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftMateriaalNaam)**
   *
   * De property heeftMateriaalNaam legt de relatie tussen het materiaal en de naam van het materiaal vast.
   *
   * @type        property
   * @domain      ceo:Materiaal
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftMateriaalNaam = this.toIri().concat('heeftMateriaalNaam')

  /**
   * **[ceo:Heeft monument aard](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftMonumentAard)**
   *
   * De property heeftMonumentAard legt de relatie tussen een rijksmonument en de aard van het monument vast, geeft aan of het rijksmonument archeologisch of onroerend gebouwd is.
   *
   * @type        property
   * @domain      ceo:Rijksmonument
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftMonumentAard = this.toIri().concat('heeftMonumentAard')

  /**
   * **[ceo:Heeft naam](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftNaam)**
   *
   * De property heeftNaam legt de relatie tussen een cultuurhistorisch object en de naam vast.
   *
   * @type        property
   * @domain      ceo:CultuurhistorischObject
   * @range       ceo:Naam
   */
  public readonly heeftNaam = this.toIri().concat('heeftNaam')

  /**
   * **[ceo:Heeft observatiedatum](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftObservatiedatum)**
   *
   * De property heeftObservatiedatum legt de relatie tussen de Kennisregistratie en de observatiedatum van de registratie vast.
   *
   * @type        property
   * @domain      ceo:Kennisregistratie
   * @range       ceo:Datering
   */
  public readonly heeftObservatiedatum = this.toIri().concat('heeftObservatiedatum')

  /**
   * **[ceo:Heeft omschrijving](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftOmschrijving)**
   *
   * De property heeftOmschrijving legt de relatie tussen een cultuurhistorisch object en de omschrijving vast.
   *
   * @type        property
   * @domain      ceo:CultuurhistorischObject
   * @range       ceo:Omschrijving
   */
  public readonly heeftOmschrijving = this.toIri().concat('heeftOmschrijving')

  /**
   * **[ceo:Heeft omschrijvingstype](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftOmschrijvingstype)**
   *
   * De property heeftOmschrijvingstype legt de relatie tussen de omschrijving en de type omschrijving met betrekking tot een cultuurhistorisch object vast.
   *
   * @type        property
   * @domain      ceo:Omschrijving
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftOmschrijvingstype = this.toIri().concat('heeftOmschrijvingstype')

  /**
   * **[ceo:Heeft oorspronkelijke functie](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftOorspronkelijkeFunctie)**
   *
   * De property heeftOorspronkelijkeFunctie legt de relatie tussen een cultuurhistorisch object en de oorspronkelijke functie vast.
   *
   * @type        property
   * @parent      ceo:heeftFunctie
   * @domain      ceo:Rijksmonument
   * @range       ceo:Functie
   */
  public readonly heeftOorspronkelijkeFunctie = this.toIri().concat('heeftOorspronkelijkeFunctie')

  /**
   * **[ceo:Heeft pand](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftPand)**
   *
   * De property heeftPand verwijst naar pandidentificatie uit de Basisregistratie Adressen en Gebouwen.
   *
   * @type        property
   * @domain      ceo:BAGRelatie
   */
  public readonly heeftPand = this.toIri().concat('heeftPand')

  /**
   * **[ceo:Heeft periode](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftPeriode)**
   *
   * De property heeftPeriode legt de relatie tussen de datering en de periode vast.
   *
   * @type        property
   * @domain      ceo:BeginDatering or ceo:EindDatering
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftPeriode = this.toIri().concat('heeftPeriode')

  /**
   * **[ceo:Heeft provincie](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftProvincie)**
   *
   * De property ceo:heeftProvincie is een registergegeven binnen de class Rijksmonument.
   *
   * @type        property
   * @domain      ceo:BasisregistratieRelatie
   * @range       http://standaarden.overheid.nl/owms/terms/Provincie
   */
  public readonly heeftProvincie = this.toIri().concat('heeftProvincie')

  /**
   * **[ceo:Heeft Rijksmonument](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftRijksmonument)**
   *
   * De property heeftSamenhang legt de relatie tussen het complex met betrekking tot een rijksmonument vast.
   *
   * @type        property
   * @domain      ceo:Complex
   * @range       ceo:Rijksmonument
   */
  public readonly heeftRijksmonument = this.toIri().concat('heeftRijksmonument')

  /**
   * **[ceo:Heeft rol](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftRol)**
   *
   * De property heeftRol legt de relatie tussen de actor en rol en de rol met betrekking tot een cultuurhistorisch object vast.
   *
   * @type        property
   * @domain      ceo:ActorEnRol
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftRol = this.toIri().concat('heeftRol')

  /**
   * **[ceo:Heeft situering](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftSituering)**
   *
   * De property heeftSituering legt de relatie tussen een rijksmonument dat geen verblijfsobject in de BAG is met een verblijfsobject in de BAG.
   *
   * @type        property
   * @domain      ceo:BAGRelatie
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftSituering = this.toIri().concat('heeftSituering')

  /**
   * **[ceo:Heeft stijl en cultuur](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftStijlEnCultuur)**
   *
   * De property heeftStijlEnCultuur legt de relatie tussen een cultuurhistorisch object en de stijl en cultuur vast.
   *
   * @type        property
   * @domain      ceo:CultuurhistorischObject
   * @range       ceo:StijlEnCultuur
   */
  public readonly heeftStijlEnCultuur = this.toIri().concat('heeftStijlEnCultuur')

  /**
   * **[ceo:Heeft stijl en cultuur naam](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftStijlEnCultuurNaam)**
   *
   * De property heeftStijlEnCultuurNaam legt de relatie tussen de stijl en cultuur en de naam van de stijl en cultuur vast.
   *
   * @type        property
   * @domain      ceo:StijlEnCultuur
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftStijlEnCultuurNaam = this.toIri().concat('heeftStijlEnCultuurNaam')

  /**
   * **[ceo:Heeft temporele relatie](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftTemporeleRelatie)**
   *
   * De property heeftTemporeleRelatie legt de relatie tussen het ... met betrekking tot een cultuurhistorisch object vast.
   *
   * @type        property
   * @domain      ceo:Datering
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftTemporeleRelatie = this.toIri().concat('heeftTemporeleRelatie')

  /**
   * **[ceo:Heeft toestand](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftToestand)**
   *
   * De property heeftToestand legt de relatie tussen de vondst en de toestand vast.
   *
   * @type        property
   * @domain      ceo:Vondsten
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftToestand = this.toIri().concat('heeftToestand')

  /**
   * **[ceo:Heeft type](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftType)**
   *
   * De property heeftType legt de relatie tussen een cultuurhistorisch object en het type vast.
   *
   * @type        property
   * @domain      ceo:CultuurhistorischObject
   * @range       ceo:Type
   */
  public readonly heeftType = this.toIri().concat('heeftType')

  /**
   * **[ceo:Heeft type naam](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftTypeNaam)**
   *
   * De property heeftTypenaam legt de relatie tussen het type en de naam van het type met betrekking tot een cultuurhistorisch object vast.
   *
   * @type        property
   * @domain      ceo:Type
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftTypeNaam = this.toIri().concat('heeftTypeNaam')

  /**
   * **[ceo:Heeft typekenmerk](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftTypekenmerk)**
   *
   * De property heeftTypekenmerk legt de relatie tussen het type en het typekenmerk met betrekking tot een cultuurhistorisch object vast.
   *
   * @type        property
   * @domain      ceo:Type
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftTypekenmerk = this.toIri().concat('heeftTypekenmerk')

  /**
   * **[ceo:heeft verblijfsobject](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftVerblijfsobject)**
   *
   * De property heeftVerblijfsobject verwijst naar de verblijfsobjectidentificatie uit de Basisregistratie Adressen en Gebouwen.
   *
   * @type        property
   * @domain      ceo:BAGRelatie
   */
  public readonly heeftVerblijfsobject = this.toIri().concat('heeftVerblijfsobject')

  /**
   * **[ceo:Heeft vertrouwelijk aanduiding](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftVertrouwelijkAanduiding)**
   *
   * De property heeftVertrouwelijkAanduiding legt de relatie tussen het cultuurhistorisch object en de vertrouwelijkheid aanduiding vast.
   *
   * @type        property
   * @domain      ceo:CultuurhistorischObject or ceo:Kennisregistratie
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftVertrouwelijkAanduiding = this.toIri().concat('heeftVertrouwelijkAanduiding')

  /**
   * **[ceo:Heeft verwerving](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftVerwerving)**
   *
   * De property heeftVerwerving legt de relatie tussen de vondst of grondspoor en de verwerving vast.
   *
   * @type        property
   * @domain      ceo:Vondstlocatie
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftVerwerving = this.toIri().concat('heeftVerwerving')

  /**
   * **[ceo:Heeft werelderfgoedstatus](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftWerelderfgoedstatus)**
   *
   * De property heeftWerelderfgoedstatus legt de relatie tussen het werelderfgoed object en zijn status vast.
   *
   * @type        property
   * @domain      ceo:Werelderfgoed
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftWerelderfgoedstatus = this.toIri().concat('heeftWerelderfgoedstatus')

  /**
   * **[ceo:Heeft zaak](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftZaak)**
   *
   * De property heeftZaak legt de relatie tussen de kennisregistratie en de zaak vast.
   *
   * @type        property
   * @domain      ceo:Kennisregistratie
   * @range       https://linkeddata.cultureelerfgoed.nl/def/zo#Zaak
   */
  public readonly heeftZaak = this.toIri().concat('heeftZaak')

  /**
   * **[ceo:Heeft zekerheidsaanduiding](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftZekerheidsaanduiding)**
   *
   * De property heeftZekerheidsaanduiding legt de relatie tussen het type en de zekerheidsaanduiding met betrekking tot een cultuurhistorisch object vast.
   *
   * @type        property
   * @domain      ceo:Type
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftZekerheidsaanduiding = this.toIri().concat('heeftZekerheidsaanduiding')

  /**
   * **[ceo:Heeft zuiverheid](https://linkeddata.cultureelerfgoed.nl/def/ceo#heeftZuiverheid)**
   *
   * De property heeftZuiverheid legt de relatie tussen de stijl en cultuur en de zuiverheid met betrekking tot een cultuurhistorisch object vast.
   *
   * @type        property
   * @domain      ceo:StijlEnCultuur
   * @range       http://www.w3.org/2004/02/skos/core#Concept
   */
  public readonly heeftZuiverheid = this.toIri().concat('heeftZuiverheid')

  /**
   * **[ceo:Hoofdfunctie](https://linkeddata.cultureelerfgoed.nl/def/ceo#hoofdfunctie)**
   *
   * De property hoofdfunctie wordt gebruikt om aan te geven of een functie de hoofdfunctie van een Cultuurhistorisch Object is.
   *
   * @type        property
   * @domain      ceo:Functie
   * @range       http://www.w3.org/2001/XMLSchema#boolean
   */
  public readonly hoofdfunctie = this.toIri().concat('hoofdfunctie')

  /**
   * **[ceo:Hoort bij zaak](https://linkeddata.cultureelerfgoed.nl/def/ceo#hoortBijZaak)**
   *
   * De property hoortBijZaak legt de relatie tussen kennisregistratie en zaak vast.
   *
   * @type        property
   * @domain      ceo:Kennisregistratie
   * @range       https://linkeddata.cultureelerfgoed.nl/def/zo#Zaak
   */
  public readonly hoortBijZaak = this.toIri().concat('hoortBijZaak')

  /**
   * **[ceo:Huidige naam](https://linkeddata.cultureelerfgoed.nl/def/ceo#huidigeNaam)**
   *
   * De property huidigeNaam wordt gebruikt om aan te geven of de naam, de huidige naam is van een cultuurhistorisch object.
   *
   * @type        property
   * @domain      ceo:Naam
   * @range       http://www.w3.org/2001/XMLSchema#boolean
   */
  public readonly huidigeNaam = this.toIri().concat('huidigeNaam')

  /**
   * **[ceo:Huisnummer](https://linkeddata.cultureelerfgoed.nl/def/ceo#huisnummer)**
   *
   * De property huisnummer is een registergegeven binnen de class Rijksmonument.
   *
   * @type        property
   * @domain      ceo:BAGRelatie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly huisnummer = this.toIri().concat('huisnummer')

  /**
   * **[ceo:Indicatie archeologische monumenten kaart waardig](https://linkeddata.cultureelerfgoed.nl/def/ceo#indicatieArcheologischeMonumentenKaartWaardig)**
   *
   * De property indicatieArcheologischeMonumentenKaartWaardig wordt gebruikt om aan te geven of een archeologisch onderzoeksgebied AMK waardig is.
   *
   * @type        property
   * @domain      ceo:ArcheologischOnderzoeksgebied
   * @range       http://www.w3.org/2001/XMLSchema#boolean
   */
  public readonly indicatieArcheologischeMonumentenKaartWaardig = this.toIri().concat('indicatieArcheologischeMonumentenKaartWaardig')

  /**
   * **[ceo:Internationaal kenteken](https://linkeddata.cultureelerfgoed.nl/def/ceo#internationaalKenteken)**
   *
   * De property internationaalKenteken wordt gebruikt om aan te geven of een internationaal kenteken aan een Cultuurhistorisch object is gegeven.
   *
   * @type        property
   * @domain      ceo:CultuurhistorischObject
   * @range       http://www.w3.org/2001/XMLSchema#boolean
   */
  public readonly internationaalKenteken = this.toIri().concat('internationaalKenteken')

  /**
   * **[ceo:Is hoofdobject van complex](https://linkeddata.cultureelerfgoed.nl/def/ceo#isHoofdobjectVanComplex)**
   *
   * De property isHoofdobjectVanComplex legt de relatie tussen een rijksmonument en het hoofdobject met betrekking tot een complex vast.
   *
   * @type        property
   * @domain      ceo:Rijksmonument
   * @range       ceo:Complex
   */
  public readonly isHoofdobjectVanComplex = this.toIri().concat('isHoofdobjectVanComplex')

  /**
   * **[ceo:Is kennisregistratie van zaak](https://linkeddata.cultureelerfgoed.nl/def/ceo#isKennisregistratieVanZaak)**
   *
   * De property isKennisregistratieVanZaak legt de relatie tussen zaak en kennisregistratie vast.
   *
   * @type        property
   * @domain      https://linkeddata.cultureelerfgoed.nl/def/zo#Zaak
   * @range       ceo:Kennisregistratie
   */
  public readonly isKennisregistratieVanZaak = this.toIri().concat('isKennisregistratieVanZaak')

  /**
   * **[ceo:Is onderdeel van complex](https://linkeddata.cultureelerfgoed.nl/def/ceo#isOnderdeelVanComplex)**
   *
   * De property isOnderdeelVanComplex legt de relatie tussen het rijksmonument als onderdeel van een complex vast.
   *
   * @type        property
   * @domain      ceo:Rijksmonument
   * @range       ceo:Complex
   */
  public readonly isOnderdeelVanComplex = this.toIri().concat('isOnderdeelVanComplex')

  /**
   * **[ceo:Kaartbladcode](https://linkeddata.cultureelerfgoed.nl/def/ceo#kaartbladcode)**
   *
   * De property kaartbladcode wordt gebruikt om aan te geven welke kaartbladcode is gekoppeld aan de locatie aanduiding.
   *
   * @type        property
   * @domain      ceo:LocatieAanduiding
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly kaartbladcode = this.toIri().concat('kaartbladcode')

  /**
   * **[ceo:Kadasterdeel](https://linkeddata.cultureelerfgoed.nl/def/ceo#kadasterdeel)**
   *
   * De property kadasterdeel wordt gebruikt om aan te geven het kadasterdeel waaronder het rijksmonument bekend is.
   *
   * @type        property
   * @domain      ceo:Rijksmonument
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly kadasterdeel = this.toIri().concat('kadasterdeel')

  /**
   * **[ceo:Kadasterdeelnummer](https://linkeddata.cultureelerfgoed.nl/def/ceo#kadasterdeelnummer)**
   *
   * De property kadasterdeelnummer wordt gebruikt om aan te geven het kadasterdeelnummer waaronder het rijksmonument bekend is.
   *
   * @type        property
   * @domain      ceo:Rijksmonument
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly kadasterdeelnummer = this.toIri().concat('kadasterdeelnummer')

  /**
   * **[ceo:Kadastrale gemeentecode](https://linkeddata.cultureelerfgoed.nl/def/ceo#kadastraleGemeentecode)**
   *
   * De property kadastraleGemeentecode is een registergegeven binnen de class Rijksmonument.
   *
   * @type        property
   * @domain      ceo:BRKRelatie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly kadastraleGemeentecode = this.toIri().concat('kadastraleGemeentecode')

  /**
   * **[ceo:Kennisregistratie toelichting](https://linkeddata.cultureelerfgoed.nl/def/ceo#kennisregistratieToelichting)**
   *
   * De property kennisegistratieToelichting wordt gebruikt om aan te geven de toelichting op de kennisregistratie instance.
   *
   * @type        property
   * @domain      ceo:Kennisregistratie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly kennisregistratieToelichting = this.toIri().concat('kennisregistratieToelichting')

  /**
   * **[ceo:Kennisegistratienummer](https://linkeddata.cultureelerfgoed.nl/def/ceo#kennisregistratienummer)**
   *
   * De property kennisegistratienummer wordt gebruikt om aan te geven het unieke nummer waaronder de kennisregistratie instance bekend is.
   *
   * @type        property
   * @domain      ceo:Kennisregistratie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly kennisregistratienummer = this.toIri().concat('kennisregistratienummer')

  /**
   * **[ceo:Ligt in object](https://linkeddata.cultureelerfgoed.nl/def/ceo#ligtInObject)**
   *
   * De property ligtInObject legt de relatie tussen de ligging van een cultuurhistorisch object met betrekking tot een cultuurhistorisch object vast.
   *
   * @type        property
   * @domain      ceo:CultuurhistorischObject
   * @range       ceo:CultuurhistorischObject
   */
  public readonly ligtInObject = this.toIri().concat('ligtInObject')

  /**
   * **[ceo:Locatie aanduidingnummer](https://linkeddata.cultureelerfgoed.nl/def/ceo#locatieAanduidingnummer)**
   *
   * De property locatieAanduidingnummer wordt gebruikt om aan te geven het unieke nummer waaronder de locatie aanduiding instance bekend is.
   *
   * @type        property
   * @domain      ceo:LocatieAanduiding
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly locatieAanduidingnummer = this.toIri().concat('locatieAanduidingnummer')

  /**
   * **[ceo:Locatienaam](https://linkeddata.cultureelerfgoed.nl/def/ceo#locatienaam)**
   *
   * De property locatienaam wordt gebruikt om aan te geven welke locatie naam is gekoppeld aan de locatie aanduiding
   *
   * @type        property
   * @domain      ceo:LocatieAanduiding
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly locatienaam = this.toIri().concat('locatienaam')

  /**
   * **[ceo:Locatieomschrijving](https://linkeddata.cultureelerfgoed.nl/def/ceo#locatieomschrijving)**
   *
   * De property locatieomschrijving wordt gebruikt om aan te geven de locatie registratie van een rijksmonument, wanneer het rijksmonument geen verblijfsobject in de BAG is.
   *
   * @type        property
   * @domain      ceo:LocatieAanduiding
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly locatieomschrijving = this.toIri().concat('locatieomschrijving')

  /**
   * **[ceo:Naam](https://linkeddata.cultureelerfgoed.nl/def/ceo#naam)**
   *
   * De property naam is een registergegeven binnen de class Rijksmonument.
   *
   * @type        property
   * @domain      ceo:Naam
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly naam = this.toIri().concat('naam')

  /**
   * **[ceo:Omschrijving](https://linkeddata.cultureelerfgoed.nl/def/ceo#omschrijving)**
   *
   * De property omschrijving is een registergegeven binnen de class Rijksmonument.
   *
   * @type        property
   * @domain      ceo:Omschrijving
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly omschrijving = this.toIri().concat('omschrijving')

  /**
   * **[ceo:Openbare ruimte](https://linkeddata.cultureelerfgoed.nl/def/ceo#openbareRuimte)**
   *
   * De property openbareRuimte is een registergegeven binnen de class Rijksmonument.
   *
   * @type        property
   * @domain      ceo:BAGRelatie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly openbareRuimte = this.toIri().concat('openbareRuimte')

  /**
   * **[ceo:Pand identificatie](https://linkeddata.cultureelerfgoed.nl/def/ceo#pandIdentificatie)**
   *
   * De property pandIdentificatie wordt gebruikt om aan te geven welke pand identificatie aan de Basisregistratie Adressen en Gebouwen is gekoppeld.
   *
   * @type        property
   * @domain      ceo:BAGRelatie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly pandIdentificatie = this.toIri().concat('pandIdentificatie')

  /**
   * **[ceo:Perceelnummer](https://linkeddata.cultureelerfgoed.nl/def/ceo#perceelnummer)**
   *
   * De property perceelnummer is een registergegeven binnen de class Rijksmonument.
   *
   * @type        property
   * @domain      ceo:BRKRelatie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly perceelnummer = this.toIri().concat('perceelnummer')

  /**
   * **[ceo:Plaatscode](https://linkeddata.cultureelerfgoed.nl/def/ceo#plaatscode)**
   *
   * De property plaatsCode wordt gebruikt om aan te geven welke plaats code is gekoppeld aan het rijksmonument.
   *
   * @type        property
   * @domain      ceo:BRKRelatie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly plaatscode = this.toIri().concat('plaatscode')

  /**
   * **[ceo:Postcode](https://linkeddata.cultureelerfgoed.nl/def/ceo#postcode)**
   *
   * De property postcode is een registergegeven binnen de class Rijksmonument.
   *
   * @type        property
   * @domain      ceo:BAGRelatie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly postcode = this.toIri().concat('postcode')

  /**
   * **[ceo:Provinciecode](https://linkeddata.cultureelerfgoed.nl/def/ceo#provinciecode)**
   *
   * De property provincieCode wordt gebruikt om aan te geven welke provincie code is gekoppeld aan het rijksmonument.
   *
   * @type        property
   * @domain      ceo:BRKRelatie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly provinciecode = this.toIri().concat('provinciecode')

  /**
   * **[ceo:registergegeven](https://linkeddata.cultureelerfgoed.nl/def/ceo#registergegeven)**
   *
   * De property is alleen van toepassing binnen de class Rijksmonument.
   *
   * @type        property
   * @range       http://www.w3.org/2001/XMLSchema#boolean
   */
  public readonly registergegeven = this.toIri().concat('registergegeven')

  /**
   * **[ceo:Registratiedatum](https://linkeddata.cultureelerfgoed.nl/def/ceo#registratiedatum)**
   *
   * De property registratiedatum wordt gebruikt om aan te geven de registratiedatum van de kennisregistratie.
   *
   * @type        property
   * @range       http://www.w3.org/2001/XMLSchema#date
   */
  public readonly registratiedatum = this.toIri().concat('registratiedatum')

  /**
   * **[ceo:Rijksmonumentnummer](https://linkeddata.cultureelerfgoed.nl/def/ceo#rijksmonumentnummer)**
   *
   * De property rijksmonumentnummer wordt gebruikt om aan te geven het unieke nummer waaronder het rijksmonument bekend is in het monumentenregister.
   *
   * @type        property
   * @domain      ceo:Rijksmonument
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly rijksmonumentnummer = this.toIri().concat('rijksmonumentnummer')

  /**
   * **[ceo:sectie](https://linkeddata.cultureelerfgoed.nl/def/ceo#sectie)**
   *
   * De property sectie is een registergegeven binnen de class Rijksmonument.
   *
   * @type        property
   * @domain      ceo:BRKRelatie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly sectie = this.toIri().concat('sectie')

  /**
   * **[ceo:Verblijfsobject identificatie](https://linkeddata.cultureelerfgoed.nl/def/ceo#verblijfsobjectIdentificatie)**
   *
   * De property verblijfsobjectIdentificatie wordt gebruikt om aan te geven welke verblijfsobject identificatie aan de Basisregistratie Adressen en Gebouwen is gekoppeld.
   *
   * @type        property
   * @domain      ceo:BAGRelatie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly verblijfsobjectIdentificatie = this.toIri().concat('verblijfsobjectIdentificatie')

  /**
   * **[ceo:Volledig adres](https://linkeddata.cultureelerfgoed.nl/def/ceo#volledigAdres)**
   *
   * De property volledigAdres wordt gebruikt om aan te geven het volledige adres van een cultuurhistorisch object.
   *
   * @type        property
   * @domain      ceo:BAGRelatie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly volledigAdres = this.toIri().concat('volledigAdres')

  /**
   * **[ceo:Werelderfgoednummer](https://linkeddata.cultureelerfgoed.nl/def/ceo#werelderfgoednummer)**
   *
   * De property wereldnummer wordt gebruikt om aan te geven het unieke nummer waaronder het werelderfgoed bekend is.
   *
   * @type        property
   * @domain      ceo:Werelderfgoed
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly werelderfgoednummer = this.toIri().concat('werelderfgoednummer')

  /**
   * **[ceo:Woonplaatsnaam](https://linkeddata.cultureelerfgoed.nl/def/ceo#woonplaatsnaam)**
   *
   * De property woonplaatsnaam is een registergegeven binnen de class Rijksmonument.
   *
   * @type        property
   * @domain      ceo:BAGRelatie
   * @range       http://www.w3.org/2001/XMLSchema#string
   */
  public readonly woonplaatsnaam = this.toIri().concat('woonplaatsnaam')
}

/**
 * prefix **ceo:** <<https://linkeddata.cultureelerfgoed.nl/def/ceo#>>
 *
 * ceo
 *
 * ceo
 */
const ceo = new Vocabulary('https://linkeddata.cultureelerfgoed.nl/def/ceo#')
export default ceo
