// @triplyetl/vocabularies/src/Vocabularies/Static/index.ts
import { Iri } from '../index.js'

export { default as cc } from './cc.js'
export { default as epsg } from './epsg.js'
export { default as owms } from './owms.js'
export { default as xsd } from './xsd.js'

/**
 * Art and Architecture Thesaurus
 */
export const aat = new Iri('http://vocab.getty.edu/aat/')

/**
 * Thesaurus of Geographic Names
 */
export const tgn = new Iri('http://vocab.getty.edu/tgn/')

/**
 * Union List of Artist Names
 */
export const ulan = new Iri('http://vocab.getty.edu/ulan/')

/**
 * Basisregistratie Nederlandse Overheid - BRK
 */
export const brk = new Iri('http://brk.basisregistratie.overheid.nl/def/brk#')

/**
 * Contact: Utility concepts for everyday life
 *
 */
export const con = new Iri('http://www.w3.org/2000/10/swap/pim/contact#')
// export const dbo = new Iri('http://dbpedia.org/ontology/')

export const dbr = new Iri('http://dbpedia.org/resource/')

/**
 * Unique URIs for File Formats
 *
 * http://www.w3.org/ns/formats/Format is a unique, top-level Class URI for specific file formats.
 * Media types do not uniquely identify file formats (eg, serialization formats for various
 * abstract concepts), because not all specific cases do have distinct media types, instances
 * in this class can be used instead for proper identification.
 */
export const format = new Iri('http://www.w3.org/ns/formats/')

/**
 * GeoNames
 *
 * The GeoNames geographical database covers all countries and contains over
 * eleven million placenames that are available for download free of charge.
 */
export const geonames = new Iri('https://www.geonames.org/')

/**
 * The comprehensive classification system for the content of images.
 *
 * Providing subject access to the collections of many museums and libraries,
 * but its vocabulary covers a spectrum of visual communication that is broader
 * than that of the traditional cultural heritage domain.
 */
export const iconclass = new Iri('http://iconclass.org/')

export const lxr = new Iri('https://linkeddata.cultureelerfgoed.nl/def/lxr#')
export const mim = new Iri('http://bp4mc2.org/def/mim#')
export const oclc = new Iri('http://www.worldcat.org/oclc/')
export const orb = new Iri('http://purl.org/orb/1.0/')
export const po = new Iri('http://www.essepuntato.it/2008/12/pattern#')
export const swap = new Iri('http://www.w3.org/2000/10/swap/pim/doc#')
export const topic = new Iri('https://triplydb.com/Triply/topics/id/')
export const vs = new Iri('http://www.w3.org/2003/06/sw-vocab-status/ns#')
export const zo = new Iri('https://linkeddata.cultureelerfgoed.nl/def/zo#')
