// @triplyetl/vocabularies/lib/Languages/region.js
import Region from '../Classes/Region.js';
const region = {
    /**
     * Private use
     */
    AA: new Region('AA'),
    /**
     * Ascension Island
     */
    AC: new Region('AC'),
    /**
     * Andorra
     */
    AD: new Region('AD'),
    /**
     * United Arab Emirates
     */
    AE: new Region('AE'),
    /**
     * Afghanistan
     */
    AF: new Region('AF'),
    /**
     * Antigua and Barbuda
     */
    AG: new Region('AG'),
    /**
     * Anguilla
     */
    AI: new Region('AI'),
    /**
     * Albania
     */
    AL: new Region('AL'),
    /**
     * Armenia
     */
    AM: new Region('AM'),
    /**
     * Netherlands Antilles
     */
    AN: new Region('AN'),
    /**
     * Angola
     */
    AO: new Region('AO'),
    /**
     * Antarctica
     */
    AQ: new Region('AQ'),
    /**
     * Argentina
     */
    AR: new Region('AR'),
    /**
     * American Samoa
     */
    AS: new Region('AS'),
    /**
     * Austria
     */
    AT: new Region('AT'),
    /**
     * Australia
     */
    AU: new Region('AU'),
    /**
     * Aruba
     */
    AW: new Region('AW'),
    /**
     * Åland Islands
     */
    AX: new Region('AX'),
    /**
     * Azerbaijan
     */
    AZ: new Region('AZ'),
    /**
     * Bosnia and Herzegovina
     */
    BA: new Region('BA'),
    /**
     * Barbados
     */
    BB: new Region('BB'),
    /**
     * Bangladesh
     */
    BD: new Region('BD'),
    /**
     * Belgium
     */
    BE: new Region('BE'),
    /**
     * Burkina Faso
     */
    BF: new Region('BF'),
    /**
     * Bulgaria
     */
    BG: new Region('BG'),
    /**
     * Bahrain
     */
    BH: new Region('BH'),
    /**
     * Burundi
     */
    BI: new Region('BI'),
    /**
     * Benin
     */
    BJ: new Region('BJ'),
    /**
     * Saint Barthélemy
     */
    BL: new Region('BL'),
    /**
     * Bermuda
     */
    BM: new Region('BM'),
    /**
     * Brunei Darussalam
     */
    BN: new Region('BN'),
    /**
     * Bolivia
     */
    BO: new Region('BO'),
    /**
     * Bonaire, Sint Eustatius and Saba
     */
    BQ: new Region('BQ'),
    /**
     * Brazil
     */
    BR: new Region('BR'),
    /**
     * Bahamas
     */
    BS: new Region('BS'),
    /**
     * Bhutan
     */
    BT: new Region('BT'),
    /**
     * Burma
     */
    BU: new Region('BU'),
    /**
     * Bouvet Island
     */
    BV: new Region('BV'),
    /**
     * Botswana
     */
    BW: new Region('BW'),
    /**
     * Belarus
     */
    BY: new Region('BY'),
    /**
     * Belize
     */
    BZ: new Region('BZ'),
    /**
     * Canada
     */
    CA: new Region('CA'),
    /**
     * Cocos (Keeling) Islands
     */
    CC: new Region('CC'),
    /**
     * The Democratic Republic of the Congo
     */
    CD: new Region('CD'),
    /**
     * Central African Republic
     */
    CF: new Region('CF'),
    /**
     * Congo
     */
    CG: new Region('CG'),
    /**
     * Switzerland
     */
    CH: new Region('CH'),
    /**
     * Côte d'Ivoire
     */
    CI: new Region('CI'),
    /**
     * Cook Islands
     */
    CK: new Region('CK'),
    /**
     * Chile
     */
    CL: new Region('CL'),
    /**
     * Cameroon
     */
    CM: new Region('CM'),
    /**
     * China
     */
    CN: new Region('CN'),
    /**
     * Colombia
     */
    CO: new Region('CO'),
    /**
     * Clipperton Island
     */
    CP: new Region('CP'),
    /**
     * Costa Rica
     */
    CR: new Region('CR'),
    /**
     * Serbia and Montenegro
     */
    CS: new Region('CS'),
    /**
     * Cuba
     */
    CU: new Region('CU'),
    /**
     * Cabo Verde, Cape Verde
     */
    CV: new Region('CV'),
    /**
     * Curaçao
     */
    CW: new Region('CW'),
    /**
     * Christmas Island
     */
    CX: new Region('CX'),
    /**
     * Cyprus
     */
    CY: new Region('CY'),
    /**
     * Czechia, Czech Republic
     */
    CZ: new Region('CZ'),
    /**
     * German Democratic Republic
     */
    DD: new Region('DD'),
    /**
     * Germany
     */
    DE: new Region('DE'),
    /**
     * Diego Garcia
     */
    DG: new Region('DG'),
    /**
     * Djibouti
     */
    DJ: new Region('DJ'),
    /**
     * Denmark
     */
    DK: new Region('DK'),
    /**
     * Dominica
     */
    DM: new Region('DM'),
    /**
     * Dominican Republic
     */
    DO: new Region('DO'),
    /**
     * Algeria
     */
    DZ: new Region('DZ'),
    /**
     * Ceuta, Melilla
     */
    EA: new Region('EA'),
    /**
     * Ecuador
     */
    EC: new Region('EC'),
    /**
     * Estonia
     */
    EE: new Region('EE'),
    /**
     * Egypt
     */
    EG: new Region('EG'),
    /**
     * Western Sahara
     */
    EH: new Region('EH'),
    /**
     * Eritrea
     */
    ER: new Region('ER'),
    /**
     * Spain
     */
    ES: new Region('ES'),
    /**
     * Ethiopia
     */
    ET: new Region('ET'),
    /**
     * European Union
     */
    EU: new Region('EU'),
    /**
     * Eurozone
     */
    EZ: new Region('EZ'),
    /**
     * Finland
     */
    FI: new Region('FI'),
    /**
     * Fiji
     */
    FJ: new Region('FJ'),
    /**
     * Falkland Islands (Malvinas)
     */
    FK: new Region('FK'),
    /**
     * Federated States of Micronesia
     */
    FM: new Region('FM'),
    /**
     * Faroe Islands
     */
    FO: new Region('FO'),
    /**
     * France
     */
    FR: new Region('FR'),
    /**
     * Metropolitan France
     */
    FX: new Region('FX'),
    /**
     * Gabon
     */
    GA: new Region('GA'),
    /**
     * United Kingdom
     */
    GB: new Region('GB'),
    /**
     * Grenada
     */
    GD: new Region('GD'),
    /**
     * Georgia
     */
    GE: new Region('GE'),
    /**
     * French Guiana
     */
    GF: new Region('GF'),
    /**
     * Guernsey
     */
    GG: new Region('GG'),
    /**
     * Ghana
     */
    GH: new Region('GH'),
    /**
     * Gibraltar
     */
    GI: new Region('GI'),
    /**
     * Greenland
     */
    GL: new Region('GL'),
    /**
     * Gambia
     */
    GM: new Region('GM'),
    /**
     * Guinea
     */
    GN: new Region('GN'),
    /**
     * Guadeloupe
     */
    GP: new Region('GP'),
    /**
     * Equatorial Guinea
     */
    GQ: new Region('GQ'),
    /**
     * Greece
     */
    GR: new Region('GR'),
    /**
     * South Georgia and the South Sandwich Islands
     */
    GS: new Region('GS'),
    /**
     * Guatemala
     */
    GT: new Region('GT'),
    /**
     * Guam
     */
    GU: new Region('GU'),
    /**
     * Guinea-Bissau
     */
    GW: new Region('GW'),
    /**
     * Guyana
     */
    GY: new Region('GY'),
    /**
     * Hong Kong
     */
    HK: new Region('HK'),
    /**
     * Heard Island and McDonald Islands
     */
    HM: new Region('HM'),
    /**
     * Honduras
     */
    HN: new Region('HN'),
    /**
     * Croatia
     */
    HR: new Region('HR'),
    /**
     * Haiti
     */
    HT: new Region('HT'),
    /**
     * Hungary
     */
    HU: new Region('HU'),
    /**
     * Canary Islands
     */
    IC: new Region('IC'),
    /**
     * Indonesia
     */
    ID: new Region('ID'),
    /**
     * Ireland
     */
    IE: new Region('IE'),
    /**
     * Israel
     */
    IL: new Region('IL'),
    /**
     * Isle of Man
     */
    IM: new Region('IM'),
    /**
     * India
     */
    IN: new Region('IN'),
    /**
     * British Indian Ocean Territory
     */
    IO: new Region('IO'),
    /**
     * Iraq
     */
    IQ: new Region('IQ'),
    /**
     * Islamic Republic of Iran
     */
    IR: new Region('IR'),
    /**
     * Iceland
     */
    IS: new Region('IS'),
    /**
     * Italy
     */
    IT: new Region('IT'),
    /**
     * Jersey
     */
    JE: new Region('JE'),
    /**
     * Jamaica
     */
    JM: new Region('JM'),
    /**
     * Jordan
     */
    JO: new Region('JO'),
    /**
     * Japan
     */
    JP: new Region('JP'),
    /**
     * Kenya
     */
    KE: new Region('KE'),
    /**
     * Kyrgyzstan
     */
    KG: new Region('KG'),
    /**
     * Cambodia
     */
    KH: new Region('KH'),
    /**
     * Kiribati
     */
    KI: new Region('KI'),
    /**
     * Comoros
     */
    KM: new Region('KM'),
    /**
     * Saint Kitts and Nevis
     */
    KN: new Region('KN'),
    /**
     * Democratic People's Republic of Korea
     */
    KP: new Region('KP'),
    /**
     * Republic of Korea
     */
    KR: new Region('KR'),
    /**
     * Kuwait
     */
    KW: new Region('KW'),
    /**
     * Cayman Islands
     */
    KY: new Region('KY'),
    /**
     * Kazakhstan
     */
    KZ: new Region('KZ'),
    /**
     * Lao People's Democratic Republic
     */
    LA: new Region('LA'),
    /**
     * Lebanon
     */
    LB: new Region('LB'),
    /**
     * Saint Lucia
     */
    LC: new Region('LC'),
    /**
     * Liechtenstein
     */
    LI: new Region('LI'),
    /**
     * Sri Lanka
     */
    LK: new Region('LK'),
    /**
     * Liberia
     */
    LR: new Region('LR'),
    /**
     * Lesotho
     */
    LS: new Region('LS'),
    /**
     * Lithuania
     */
    LT: new Region('LT'),
    /**
     * Luxembourg
     */
    LU: new Region('LU'),
    /**
     * Latvia
     */
    LV: new Region('LV'),
    /**
     * Libya
     */
    LY: new Region('LY'),
    /**
     * Morocco
     */
    MA: new Region('MA'),
    /**
     * Monaco
     */
    MC: new Region('MC'),
    /**
     * Moldova
     */
    MD: new Region('MD'),
    /**
     * Montenegro
     */
    ME: new Region('ME'),
    /**
     * Saint Martin (French part)
     */
    MF: new Region('MF'),
    /**
     * Madagascar
     */
    MG: new Region('MG'),
    /**
     * Marshall Islands
     */
    MH: new Region('MH'),
    /**
     * North Macedonia
     */
    MK: new Region('MK'),
    /**
     * Mali
     */
    ML: new Region('ML'),
    /**
     * Myanmar
     */
    MM: new Region('MM'),
    /**
     * Mongolia
     */
    MN: new Region('MN'),
    /**
     * Macao
     */
    MO: new Region('MO'),
    /**
     * Northern Mariana Islands
     */
    MP: new Region('MP'),
    /**
     * Martinique
     */
    MQ: new Region('MQ'),
    /**
     * Mauritania
     */
    MR: new Region('MR'),
    /**
     * Montserrat
     */
    MS: new Region('MS'),
    /**
     * Malta
     */
    MT: new Region('MT'),
    /**
     * Mauritius
     */
    MU: new Region('MU'),
    /**
     * Maldives
     */
    MV: new Region('MV'),
    /**
     * Malawi
     */
    MW: new Region('MW'),
    /**
     * Mexico
     */
    MX: new Region('MX'),
    /**
     * Malaysia
     */
    MY: new Region('MY'),
    /**
     * Mozambique
     */
    MZ: new Region('MZ'),
    /**
     * Namibia
     */
    NA: new Region('NA'),
    /**
     * New Caledonia
     */
    NC: new Region('NC'),
    /**
     * Niger
     */
    NE: new Region('NE'),
    /**
     * Norfolk Island
     */
    NF: new Region('NF'),
    /**
     * Nigeria
     */
    NG: new Region('NG'),
    /**
     * Nicaragua
     */
    NI: new Region('NI'),
    /**
     * Netherlands
     */
    NL: new Region('NL'),
    /**
     * Norway
     */
    NO: new Region('NO'),
    /**
     * Nepal
     */
    NP: new Region('NP'),
    /**
     * Nauru
     */
    NR: new Region('NR'),
    /**
     * Neutral Zone
     */
    NT: new Region('NT'),
    /**
     * Niue
     */
    NU: new Region('NU'),
    /**
     * New Zealand
     */
    NZ: new Region('NZ'),
    /**
     * Oman
     */
    OM: new Region('OM'),
    /**
     * Panama
     */
    PA: new Region('PA'),
    /**
     * Peru
     */
    PE: new Region('PE'),
    /**
     * French Polynesia
     */
    PF: new Region('PF'),
    /**
     * Papua New Guinea
     */
    PG: new Region('PG'),
    /**
     * Philippines
     */
    PH: new Region('PH'),
    /**
     * Pakistan
     */
    PK: new Region('PK'),
    /**
     * Poland
     */
    PL: new Region('PL'),
    /**
     * Saint Pierre and Miquelon
     */
    PM: new Region('PM'),
    /**
     * Pitcairn
     */
    PN: new Region('PN'),
    /**
     * Puerto Rico
     */
    PR: new Region('PR'),
    /**
     * State of Palestine
     */
    PS: new Region('PS'),
    /**
     * Portugal
     */
    PT: new Region('PT'),
    /**
     * Palau
     */
    PW: new Region('PW'),
    /**
     * Paraguay
     */
    PY: new Region('PY'),
    /**
     * Qatar
     */
    QA: new Region('QA'),
    /**
     * Réunion
     */
    RE: new Region('RE'),
    /**
     * Romania
     */
    RO: new Region('RO'),
    /**
     * Serbia
     */
    RS: new Region('RS'),
    /**
     * Russian Federation
     */
    RU: new Region('RU'),
    /**
     * Rwanda
     */
    RW: new Region('RW'),
    /**
     * Saudi Arabia
     */
    SA: new Region('SA'),
    /**
     * Solomon Islands
     */
    SB: new Region('SB'),
    /**
     * Seychelles
     */
    SC: new Region('SC'),
    /**
     * Sudan
     */
    SD: new Region('SD'),
    /**
     * Sweden
     */
    SE: new Region('SE'),
    /**
     * Singapore
     */
    SG: new Region('SG'),
    /**
     * Saint Helena, Ascension and Tristan da Cunha
     */
    SH: new Region('SH'),
    /**
     * Slovenia
     */
    SI: new Region('SI'),
    /**
     * Svalbard and Jan Mayen
     */
    SJ: new Region('SJ'),
    /**
     * Slovakia
     */
    SK: new Region('SK'),
    /**
     * Sierra Leone
     */
    SL: new Region('SL'),
    /**
     * San Marino
     */
    SM: new Region('SM'),
    /**
     * Senegal
     */
    SN: new Region('SN'),
    /**
     * Somalia
     */
    SO: new Region('SO'),
    /**
     * Suriname
     */
    SR: new Region('SR'),
    /**
     * South Sudan
     */
    SS: new Region('SS'),
    /**
     * Sao Tome and Principe
     */
    ST: new Region('ST'),
    /**
     * Union of Soviet Socialist Republics
     */
    SU: new Region('SU'),
    /**
     * El Salvador
     */
    SV: new Region('SV'),
    /**
     * Sint Maarten (Dutch part)
     */
    SX: new Region('SX'),
    /**
     * Syrian Arab Republic
     */
    SY: new Region('SY'),
    /**
     * Eswatini, eSwatini, Swaziland
     */
    SZ: new Region('SZ'),
    /**
     * Tristan da Cunha
     */
    TA: new Region('TA'),
    /**
     * Turks and Caicos Islands
     */
    TC: new Region('TC'),
    /**
     * Chad
     */
    TD: new Region('TD'),
    /**
     * French Southern Territories
     */
    TF: new Region('TF'),
    /**
     * Togo
     */
    TG: new Region('TG'),
    /**
     * Thailand
     */
    TH: new Region('TH'),
    /**
     * Tajikistan
     */
    TJ: new Region('TJ'),
    /**
     * Tokelau
     */
    TK: new Region('TK'),
    /**
     * Timor-Leste
     */
    TL: new Region('TL'),
    /**
     * Turkmenistan
     */
    TM: new Region('TM'),
    /**
     * Tunisia
     */
    TN: new Region('TN'),
    /**
     * Tonga
     */
    TO: new Region('TO'),
    /**
     * East Timor
     */
    TP: new Region('TP'),
    /**
     * Turkey
     */
    TR: new Region('TR'),
    /**
     * Trinidad and Tobago
     */
    TT: new Region('TT'),
    /**
     * Tuvalu
     */
    TV: new Region('TV'),
    /**
     * Taiwan, Province of China
     */
    TW: new Region('TW'),
    /**
     * United Republic of Tanzania
     */
    TZ: new Region('TZ'),
    /**
     * Ukraine
     */
    UA: new Region('UA'),
    /**
     * Uganda
     */
    UG: new Region('UG'),
    /**
     * United States Minor Outlying Islands
     */
    UM: new Region('UM'),
    /**
     * United Nations
     */
    UN: new Region('UN'),
    /**
     * United States
     */
    US: new Region('US'),
    /**
     * Uruguay
     */
    UY: new Region('UY'),
    /**
     * Uzbekistan
     */
    UZ: new Region('UZ'),
    /**
     * Holy See (Vatican City State)
     */
    VA: new Region('VA'),
    /**
     * Saint Vincent and the Grenadines
     */
    VC: new Region('VC'),
    /**
     * Venezuela
     */
    VE: new Region('VE'),
    /**
     * British Virgin Islands
     */
    VG: new Region('VG'),
    /**
     * U.S. Virgin Islands
     */
    VI: new Region('VI'),
    /**
     * Viet Nam
     */
    VN: new Region('VN'),
    /**
     * Vanuatu
     */
    VU: new Region('VU'),
    /**
     * Wallis and Futuna
     */
    WF: new Region('WF'),
    /**
     * Samoa
     */
    WS: new Region('WS'),
    /**
     * Democratic Yemen
     */
    YD: new Region('YD'),
    /**
     * Yemen
     */
    YE: new Region('YE'),
    /**
     * Mayotte
     */
    YT: new Region('YT'),
    /**
     * Yugoslavia
     */
    YU: new Region('YU'),
    /**
     * South Africa
     */
    ZA: new Region('ZA'),
    /**
     * Zambia
     */
    ZM: new Region('ZM'),
    /**
     * Zaire
     */
    ZR: new Region('ZR'),
    /**
     * Zimbabwe
     */
    ZW: new Region('ZW'),
    /**
     * Private use
     */
    ZZ: new Region('ZZ')
};
export default region;
