// @triplyetl/vocabularies/src/Commands/remove.ts
import { type CommandModule } from 'yargs'
import getJobs from '../Utils/getJobs.js'
import chalk from 'chalk'
import writeJobs from '../Utils/writeJobs.js'
import { existsSync, unlinkSync } from 'fs'
import path from 'path'

const $jobs = getJobs()

export const remove: CommandModule = {
  command: 'remove <prefixes...>',
  describe: 'remove one ore more existing vocabulary jobs',
  builder: {
    prefixes: {
      type: 'string',
      hidden: true,
      choices: $jobs.map((job) => job.prefix),
      describe: 'List of prefixes you want to delete.'
    }
  },
  handler: async (argv) => {
    const prefixes = argv.prefixes as string[]
    const jobs = $jobs.filter((job) => !prefixes.includes(job.prefix))
    writeJobs(jobs)
    for (const prefix of prefixes) {
      const file = path.join(process.cwd(), 'src', 'Vocabularies', 'Generated', `${prefix}.ts`)
      if (existsSync(file)) unlinkSync(file)
    }
    process.stdout.write(chalk.blue(`You have succesfully delete one or more jobs with prefixes '${chalk.bold(`${prefixes.join(' | ')}`)}'.\n`))
    process.stdout.write(chalk.cyan('Tip: run `npx vocabs generate --index` to regenerate the correct index files.\n'))
  }
}
