// @triplyetl/etl/lib/ratt/middlewares/transformations/index.js
// /**
//  * The RATT transformation functions.
//  *
//  * Every transformation function takes zero or more values from the current
//  * Etl Record, performs some transformation, and then stores the result into
//  * a new key.
//  */
export { default as copy } from "./copy.js";
export { default as concat } from "./concat.js";
export { default as decodeHtml } from "./decodeHtml.js";
export { default as geojsonToWkt } from "./geojsonToWkt.js";
export { default as addSkolemIri } from "./addSkolemIri.js";
export { default as addHashedIri } from "./addHashedIri.js";
export { default as addRandomIri } from "./addRandomIri.js";
export { default as addTag } from "./addTag.js";
export { default as addValue } from "./addValue.js";
export { default as addIri } from "./addIri.js";
export { default as addLiteral } from "./addLiteral.js";
export { default as lowercase } from "./lowercase.js";
export { default as replace } from "./replace.js";
export { default as split } from "./split.js";
export { default as substring } from "./substring.js";
export { default as translateAll } from "./translateAll.js";
export { default as translateSome } from "./translateSome.js";
export { default as tryLiteral } from "./tryLiteral.js";
export { default as uppercase } from "./uppercase.js";
export { default as capitalize } from "./capitalize.js";
export { default as padStart } from "./padStart.js";
export { default as padEnd } from "./padEnd.js";
export * as wkt from "./wkt/index.js";
export * as custom from "./custom/index.js";
