// @triplydb/utils/src/GeoArea.ts
import { geoProject } from "./GeoProject.js";
import type { Geometry } from "@triplydb/recognized-datatypes/wkt.js";
// @ts-ignore Typescript issue... Should be fixed in v7
import area from "@turf/area";
// @ts-ignore Typescript issue... Should be fixed in v7
import { convertArea } from "@turf/helpers";
import { isSupportedCrs } from "./Crs.js";

function prefixer<Prefix extends string>(base: Prefix) {
  return function prefix<Suffix extends string>(local: Suffix): `${Prefix}${Suffix}` {
    return `${base}${local}`;
  };
}

export const uom = prefixer("http://www.opengis.net/def/uom/OGC/1.0/");
export const unit = prefixer("http://qudt.org/vocab/unit/");

// Use meters-squared as the default unit for area calculations.
export const DEFAULT_UNIT = unit("M2");
// Required by the Turf library as the default CRS for geometries.
const CRS84 = "http://www.opengis.net/def/crs/OGC/1.3/CRS84";

const UNIT_DEF = [uom("metre"), uom("meter"), unit("IN2"), unit("M2"), unit("MI2"), unit("YD2")];
export type UnitDefType = (typeof UNIT_DEF)[number];

/**
 * Check whether unit we're converting to is support
 */
export function isSupportedAreaUnit(unit: string): unit is UnitDefType {
  return (UNIT_DEF as string[]).includes(unit);
}
/**
 * Implementation of GeoSPARQL 1.1 area function.
 * See here details https://opengeospatial.github.io/ogc-geosparql/geosparql11/spec.html#_function_geofarea
 * @param geometry Any geometry.
 * @param toUnit The unit of length/area to convert the area to. Units of length are taken to mean the unit of the
 *               length of sides of the polygon.
 * @returns 0 for any non-polygon geometry. Otherwise, returns the area in the unit specified, as a number.
 */
export function geoArea(geometry: Geometry, toUnit: UnitDefType = DEFAULT_UNIT) {
  if (!isSupportedCrs(geometry.crs)) {
    throw new Error(`Calculating the area of CRS ${geometry.crs} is not currently supported`);
  }
  if (!isSupportedAreaUnit(toUnit)) {
    throw new Error(`Calculating the area in unit ${toUnit} is not currently supported`);
  }

  // According to the spec, the area of any non-polygon geometry is 0. Turf doesn't support Triangles at this point,
  // so we don't include them either. If there is demand for this, we can add it later (as a polygon).
  if (geometry.type !== "Polygon" && geometry.type !== "MultiPolygon") return 0;

  // Transform to CRS84. This is the expected CRS of the area function.
  if (geometry.crs !== CRS84) geometry = geoProject(geometry, CRS84);

  const areaInSquareMeters = area(geometry);

  // Convert to the target units
  switch (toUnit) {
    case uom("meter"):
    case uom("metre"):
    case unit("M2"):
      return areaInSquareMeters;
    case unit("IN2"):
      return convertArea(areaInSquareMeters, "meters", "inches");
    case unit("MI2"):
      return convertArea(areaInSquareMeters, "meters", "miles");
    case unit("YD2"):
      return convertArea(areaInSquareMeters, "meters", "yards");
  }
}
