// @triplyetl/etl/lib/bin/scripts/commands/deleteQueries.js
import App from "@triply/triplydb/App.js";
import chalk from "chalk";
import inquirer from "inquirer";
import { Etl } from "../../../generic/index.js";
const cmd = {
    command: "delete-queries",
    describe: "Bulk deleting of Queries",
    builder: {},
    handler: async (_argv) => {
        if (!Etl.token) {
            console.warn(chalk.bgRed("No TriplyDB Token found, create one with the `create-token` command"));
            process.exit(1);
        }
        const app = App.get(Etl.token);
        const accounts = [];
        for await (const account of app.getAccounts()) {
            const info = await account.getInfo();
            accounts.push({
                name: `${info.name ?? info.accountName} ${info.type === "org" ? "(org)" : ""}`,
                value: info.accountName,
            });
        }
        const questionsAccount = [
            {
                name: "accountName",
                message: "Choose the account containing the queries you want to delete",
                type: "list",
                choices: accounts,
            },
        ];
        let defaultAnswerAccounts;
        if (accounts.length === 1) {
            defaultAnswerAccounts = { dataset: accounts[0].value };
        }
        const accountName = (await inquirer.prompt(questionsAccount, defaultAnswerAccounts)).accountName;
        const account = await app.getAccount(accountName);
        const foundQueries = [];
        for await (const query of account.getQueries()) {
            const info = await query.getInfo();
            foundQueries.push({
                value: info.name,
                name: `${info.description === "" ? info.name : info.description} • ${info.numberOfVersions} version(s)`,
            });
        }
        if (foundQueries.length === 0) {
            console.warn(chalk.bgMagenta("No queries found"));
            process.exit(0);
        }
        const questions = [
            {
                name: "queries",
                message: "Choose the queries you want to delete",
                type: "checkbox",
                choices: foundQueries,
            },
        ];
        inquirer
            .prompt(questions)
            .then((answers) => {
            const queries = answers.queries;
            queries.forEach((name) => {
                account
                    .getQuery(name)
                    .then(async (query) => query.delete())
                    .then((_) => console.info(chalk.green(`query '${name}' deleted`)))
                    .catch((e) => console.error(e.message.bgRed));
            });
        })
            .catch((e) => console.error(e.message.bgRed));
    },
};
export default cmd;
