// @triplyetl/etl/lib/generic/Record.js
export const key = {
    /**
     * The name of the source (e.g. TriplyDB asset) that contains the current
     * Etl record.
     */
    fileName: "$fileName",
    /**
     * The mode in which the etl is run.  See the `Mode` type.
     */
    environment: "$environment",
    /**
     * For list-shaped source data: a link to the context that directly
     * contains the list in which the current Etl record appears as a element.
     */
    parent: "$parent",
    /**
     * The numeric identifier of the current Etl record.
     */
    recordId: "$recordId",
    /**
     * For tree-shaped source data: the full tree object.
     */
    root: "$root",
    /**
     * For tree-shaped source data: the path through which the current Etl
     * record was selected.
     */
    selector: "$selector",
    /**
     * For tree-shaped source data: the last tag in the path through which the
     * current Etl record was selected.  This tag often indicates what kind of
     * record is obtained.
     */
    tagName: "$tagName",
    /**
     * For Xlsx (Excel) source data: thename of the sheet from which the
     * current Etl record was selected.
     */
    sheetName: "$sheetName",
};
